<!DOCTYPE html>
<html class="writer-html5" lang="python" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>GenAIRR.sequence.sequence &mdash; GenAIRR  documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=92fd9be5" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../../_static/documentation_options.js?v=d048f138"></script>
        <script src="../../../_static/doctools.js?v=888ff710"></script>
        <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            GenAIRR
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html">src</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">GenAIRR</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">GenAIRR.sequence.sequence</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for GenAIRR.sequence.sequence</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">..alleles</span> <span class="kn">import</span> <span class="n">AlleleTypes</span>
<span class="kn">from</span> <span class="nn">abc</span> <span class="kn">import</span> <span class="n">ABC</span><span class="p">,</span> <span class="n">abstractmethod</span>

<span class="kn">from</span> <span class="nn">..utilities.data_config</span> <span class="kn">import</span> <span class="n">DataConfig</span>


<div class="viewcode-block" id="BaseSequence">
<a class="viewcode-back" href="../../../GenAIRR.sequence.html#GenAIRR.sequence.sequence.BaseSequence">[docs]</a>
<span class="k">class</span> <span class="nc">BaseSequence</span><span class="p">(</span><span class="n">ABC</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Abstract base class representing a recombined Immunoglobulin (Ig) sequence,</span>
<span class="sd">    consisting of variable (V), diversity (D), and joining (J) gene segments.</span>

<span class="sd">    This class provides a template for creating and manipulating Ig sequences, including</span>
<span class="sd">    features like (NP) regions, junctions, and mutations.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        v_allele (Allele): The V gene allele.</span>
<span class="sd">        d_allele (Allele): The D gene allele, if present.</span>
<span class="sd">        j_allele (Allele): The J gene allele.</span>
<span class="sd">        NP1_region (str): The nucleotide sequence of the NP1 region between V and D genes.</span>
<span class="sd">        NP1_length (int): The length of the NP1 region.</span>
<span class="sd">        NP2_region (str): The nucleotide sequence of the NP2 region between D and J genes.</span>
<span class="sd">        NP2_length (int): The length of the NP2 region.</span>
<span class="sd">        ungapped_seq (str): The complete ungapped nucleotide sequence.</span>
<span class="sd">        gapped_seq (str): The complete nucleotide sequence with gaps.</span>
<span class="sd">        mutated_seq (str): The ungapped nucleotide sequence after applying mutations.</span>
<span class="sd">        gapped_mutated_seq (str): The gapped nucleotide sequence after applying mutations.</span>
<span class="sd">        junction (str): The nucleotide sequence of the junction region (CDR3 plus anchors).</span>
<span class="sd">        v_seq_start (int): The start position of the V segment in the sequence.</span>
<span class="sd">        d_seq_start (int): The start position of the D segment in the sequence, if present.</span>
<span class="sd">        j_seq_start (int): The start position of the J segment in the sequence.</span>
<span class="sd">        v_seq_end (int): The end position of the V segment in the sequence.</span>
<span class="sd">        d_seq_end (int): The end position of the D segment in the sequence, if present.</span>
<span class="sd">        j_seq_end (int): The end position of the J segment in the sequence.</span>
<span class="sd">        mutations (str): A string representation of mutation events.</span>
<span class="sd">        mutation_count (int): The total count of mutations.</span>
<span class="sd">        mutation_freq (float): The frequency of mutations in the sequence.</span>

<span class="sd">    Args:</span>
<span class="sd">        alleles (List[Allele]): A list of  V, D (optional), and J gene alleles.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">alleles</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initializes a BaseSequence instance with given alleles for V, D (optional), and J segments.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">v_allele</span> <span class="o">=</span> <span class="nb">next</span><span class="p">((</span><span class="n">allele</span> <span class="k">for</span> <span class="n">allele</span> <span class="ow">in</span> <span class="n">alleles</span> <span class="k">if</span> <span class="n">allele</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="n">AlleleTypes</span><span class="o">.</span><span class="n">V</span><span class="p">),</span> <span class="kc">None</span><span class="p">)</span>
        <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">v_allele</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>  <span class="c1"># Must Have V Allele!</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">d_allele</span> <span class="o">=</span> <span class="nb">next</span><span class="p">((</span><span class="n">allele</span> <span class="k">for</span> <span class="n">allele</span> <span class="ow">in</span> <span class="n">alleles</span> <span class="k">if</span> <span class="n">allele</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="n">AlleleTypes</span><span class="o">.</span><span class="n">D</span><span class="p">),</span> <span class="kc">None</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">j_allele</span> <span class="o">=</span> <span class="nb">next</span><span class="p">((</span><span class="n">allele</span> <span class="k">for</span> <span class="n">allele</span> <span class="ow">in</span> <span class="n">alleles</span> <span class="k">if</span> <span class="n">allele</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="n">AlleleTypes</span><span class="o">.</span><span class="n">J</span><span class="p">),</span> <span class="kc">None</span><span class="p">)</span>
        <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">j_allele</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>  <span class="c1"># Must Have J Allele!</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">NP1_region</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">NP2_region</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">NP1_length</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">NP2_length</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">junction</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">v_seq_start</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">d_seq_start</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">j_seq_start</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">v_seq_end</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">d_seq_end</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">j_seq_end</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mutation_count</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mutation_freq</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">functional</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">seq_stop_codon</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ungapped_seq</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mutated_seq</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gapped_seq</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gapped_mutated_seq</span> <span class="o">=</span> <span class="kc">None</span>

<div class="viewcode-block" id="BaseSequence.simulate_sequence">
<a class="viewcode-back" href="../../../GenAIRR.sequence.html#GenAIRR.sequence.sequence.BaseSequence.simulate_sequence">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">simulate_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">dataconfig</span><span class="p">:</span> <span class="n">DataConfig</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Abstract method to simulate the recombined nucleotide sequence including V(D)J recombination,</span>
<span class="sd">        trimming, and NP region addition based on provided data configurations.</span>

<span class="sd">        Args:</span>
<span class="sd">            dataconfig (DataConfig): An instance of DataConfig containing all necessary parameters</span>
<span class="sd">            and configurations for sequence simulation.</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: The simulated nucleotide sequence, optionally including gaps.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">pass</span></div>


<div class="viewcode-block" id="BaseSequence.get_junction_length">
<a class="viewcode-back" href="../../../GenAIRR.sequence.html#GenAIRR.sequence.sequence.BaseSequence.get_junction_length">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">get_junction_length</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Abstract method to calculate the length of the junction region of the sequence.</span>
<span class="sd">        The junction region typically includes the CDR3 region along with conserved residues</span>
<span class="sd">        from V and J gene segments.</span>

<span class="sd">        Returns:</span>
<span class="sd">            int: The length of the junction region in nucleotides.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">pass</span></div>
</div>




</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Thomas Konstantinovsky &amp; Ayelet Peres.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>