{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generating a Uniform Allele Distribution Dataset\n",
    "\n",
    "This notebook outlines the process for generating synthetic Adaptive Immune Receptor Repertoire (AIRR) sequences using GenAIRR. The focus is on leveraging GenAIRR's modular pipeline structure to create a dataset with uniformly distributed allele combinations. The dataset can be used for benchmarking and analyzing various bioinformatics tools and methods.\n",
    "\n",
    "## Initial Setup and Imports\n",
    "\n",
    "In this section, we import the necessary modules from GenAIRR and related Python libraries. This includes:\n",
    "- Core GenAIRR classes for constructing simulation pipelines.\n",
    "- Steps for sequence simulation and various augmentation processes.\n",
    "- Mutation models and data configurations.\n"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-21T13:19:05.100470Z",
     "start_time": "2025-05-21T13:19:04.264736Z"
    }
   },
   "source": [
    "# Importing GenAIRR classes\n",
    "from GenAIRR.pipeline import AugmentationPipeline\n",
    "from GenAIRR.steps import SimulateSequence,FixVPositionAfterTrimmingIndexAmbiguity,FixDPositionAfterTrimmingIndexAmbiguity,FixJPositionAfterTrimmingIndexAmbiguity\n",
    "from GenAIRR.steps import CorrectForVEndCut,CorrectForDTrims,CorruptSequenceBeginning,InsertNs,InsertIndels,ShortDValidation,DistillMutationRate\n",
    "from GenAIRR.mutation import S5F\n",
    "from GenAIRR.steps.StepBase import AugmentationStep\n",
    "from GenAIRR.parameters import ChainType,CHAIN_TYPE_INFO\n",
    "from GenAIRR.data import builtin_heavy_chain_data_config\n",
    "import pandas as pd\n",
    "import random\n",
    "from itertools import product\n",
    "# Initialize DataConfig with the path to your configuration\n",
    "#data_config = DataConfig('/path/to/your/config')\n",
    "# Or Use one of Our Builtin Data Configs\n",
    "data_config_builtin = builtin_heavy_chain_data_config()\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": 1
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Preparing Allele Lists\n",
    "We extract all possible alleles from the DataConfig to create lists of V, D, and J alleles. This helps in forming combinations for the dataset."
   ]
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-21T13:19:06.241514Z",
     "start_time": "2025-05-21T13:19:06.229641Z"
    }
   },
   "source": [
    "V_alleles = [i for j in data_config_builtin.v_alleles for i in data_config_builtin.v_alleles[j]]\n",
    "D_alleles = [i for j in data_config_builtin.d_alleles for i in data_config_builtin.d_alleles[j]]\n",
    "J_alleles = [i for j in data_config_builtin.j_alleles for i in data_config_builtin.j_alleles[j]]\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": 2
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generating Allele Combinations\n",
    "\n",
    "We use Python's itertools.product to create all possible combinations of V-D, D-J, and V-D-J alleles.\n",
    "\n",
    "This approach ensures that each possible combination of alleles is represented in the dataset, allowing for comprehensive benchmarking of tools and analyses that require diverse sequence inputs."
   ]
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-21T13:19:07.862455Z",
     "start_time": "2025-05-21T13:19:07.850346Z"
    }
   },
   "source": [
    "# Repeat each allele K times\n",
    "V_alleles = V_alleles\n",
    "D_alleles = D_alleles\n",
    "J_alleles = J_alleles\n",
    "\n",
    "\n",
    "# Shuffle each list\n",
    "random.shuffle(V_alleles)\n",
    "random.shuffle(D_alleles)\n",
    "random.shuffle(J_alleles)\n",
    "\n",
    "# Generate uniform combinations\n",
    "V_D_combinations = list(product(V_alleles, D_alleles))\n",
    "D_J_combinations = list(product(D_alleles, J_alleles))\n",
    "V_D_J_combinations = list(product(V_alleles, D_alleles,J_alleles))\n",
    "\n",
    "all_combinations = V_D_combinations+D_J_combinations+V_D_J_combinations\n",
    "\n",
    "total_combinations = len(V_D_combinations) + len(D_J_combinations) + len(V_D_J_combinations)\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": 3
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-21T13:19:08.864750Z",
     "start_time": "2025-05-21T13:19:08.853710Z"
    }
   },
   "source": [
    "print(f\"Number of V-D combinations: {len(V_D_combinations):,}\")\n",
    "print(f\"Number of D-J combinations: {len(D_J_combinations):,}\")\n",
    "print(f\"Number of V-D-J combinations: {len(V_D_J_combinations):,}\")\n",
    "print(f\"Total number of combinations: {len(V_D_combinations) + len(D_J_combinations) + len(V_D_J_combinations):,}\")"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of V-D combinations: 6,534\n",
      "Number of D-J combinations: 231\n",
      "Number of V-D-J combinations: 45,738\n",
      "Total number of combinations: 52,503\n"
     ]
    }
   ],
   "execution_count": 4
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Setting Up the Augmentation Pipeline\n",
    "In this step, we set up a GenAIRR simulation pipeline, leveraging the built-in BCR HeavyChain pipeline as the base structure. The objective is to iterate over the list of allele combinations generated earlier and dynamically update the first step of the pipeline to enforce specific V/D/J alleles for each sequence. This approach ensures that each generated sequence adheres to predefined allele combinations, maintaining control and consistency in the simulated dataset."
   ]
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-21T13:20:04.487743Z",
     "start_time": "2025-05-21T13:19:50.936776Z"
    }
   },
   "source": [
    "from tqdm.auto import tqdm\n",
    "from GenAIRR.mutation import Uniform # as an example we will use a Uniform Mutation Model\n",
    "# Set the dataconfig and the chain type for the simulations\n",
    "AugmentationStep.set_dataconfig(config=CHAIN_TYPE_INFO[ChainType.BCR_HEAVY].dataconfig,chain_type=ChainType.BCR_HEAVY)\n",
    "\n",
    "pipeline_steps = [\n",
    "    SimulateSequence(mutation_model = Uniform(min_mutation_rate=0.003,max_mutation_rate=0.25)),\n",
    "    FixVPositionAfterTrimmingIndexAmbiguity(),\n",
    "    FixDPositionAfterTrimmingIndexAmbiguity(),\n",
    "    FixJPositionAfterTrimmingIndexAmbiguity(),\n",
    "    CorrectForVEndCut(),\n",
    "    CorrectForDTrims(),\n",
    "    CorruptSequenceBeginning(corruption_probability = 0.7,corrupt_events_proba = [0.4,0.4,0.2],max_sequence_length = 576,nucleotide_add_coefficient = 210,\n",
    "                             nucleotide_remove_coefficient = 310,nucleotide_add_after_remove_coefficient = 50,random_sequence_add_proba = 1,\n",
    "                             single_base_stream_proba = 0,duplicate_leading_proba = 0,random_allele_proba = 0),\n",
    "    InsertNs(n_ratio = 0.02,proba = 0.5),\n",
    "    ShortDValidation(short_d_length= 5),\n",
    "    InsertIndels(indel_probability = 0.5,max_indels = 5,insertion_proba=0.5,deletion_proba=0.5),\n",
    "    DistillMutationRate()\n",
    "    ]\n",
    "\n",
    "\n",
    "simulated_sequences = []\n",
    "n_repeats = 5\n",
    "\n",
    "# Below can be replaced / rpeated for any specific combinations\n",
    "for V,D in tqdm(V_D_combinations):\n",
    "        sim_step = SimulateSequence(mutation_model = Uniform(min_mutation_rate=0.003,max_mutation_rate=0.25),\n",
    "                         productive = False,\n",
    "                         specific_v=V,\n",
    "                         specific_d=D)\n",
    "        pipeline_steps[0] = sim_step\n",
    "        # set the above sequence simulation step as the first step in the pipeline\n",
    "        pipeline = AugmentationPipeline(pipeline_steps)\n",
    "        for _ in range(n_repeats): # generate n_repeats copies of the same combinations\n",
    "            simulated_sequences.append(pipeline.execute().get_dict())\n",
    "\n"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "  0%|          | 0/6534 [00:00<?, ?it/s]"
      ],
      "application/vnd.jupyter.widget-view+json": {
       "version_major": 2,
       "version_minor": 0,
       "model_id": "c81d2007e3804c798e8b9efd3b6414ef"
      }
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "execution_count": 7
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-21T13:20:04.518362Z",
     "start_time": "2025-05-21T13:20:04.503810Z"
    }
   },
   "source": [
    "print(f\"Simulated Sequences : {len(simulated_sequences):,}\")"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Simulated Sequences : 32,670\n"
     ]
    }
   ],
   "execution_count": 8
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-21T13:20:04.838633Z",
     "start_time": "2025-05-21T13:20:04.556261Z"
    }
   },
   "source": [
    "pd.DataFrame(simulated_sequences)"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                                                sequence  \\\n",
       "0      CAGATGCAGCTCCAGGAGTCAAGCTCACGGCTGGCCTGGCCTTTAC...   \n",
       "1      CAGCTGCAGCTGCAGAGAGTCCGGCTCAGGACTGGTGAAGCCTTCA...   \n",
       "2      AGTCATCGGTAAAGGCCCTTGAGGGGAGAGGGTGCATCAATCCTAA...   \n",
       "3      CAGCTGCAGCTGCGTGAGTCCGGCTCAGGACTGGTGAAGCCTTCAC...   \n",
       "4      TGGAGCTGGATCCGGAGGTCAGAAGGGAAGGGCCTGGAGAGGAGAG...   \n",
       "...                                                  ...   \n",
       "32665  CATGTGCAGCTGCATGAGTCGGGCCCTGAACTGTCTAGGCCTTCGC...   \n",
       "32666  CTACGGATTTTCACCGTCACCCACGTTTCAAAATGCTCCGTCGCAT...   \n",
       "32667  CTCCCTCAAGAGTCGAGTCATCATATCAGTTGACCCGTCTAAGAAC...   \n",
       "32668  CAGGTGCAGCTTCAGGAGNCGGGCCCAGGACCGGTGATGCGCTCTG...   \n",
       "32669  CAGNTGAAGGTGCAGTAGTCGGGGCCGGACTGGTGAAGCCTTGGGA...   \n",
       "\n",
       "                                                  v_call  \\\n",
       "0                                        [IGHVF3-G12*04]   \n",
       "1                                        [IGHVF3-G12*04]   \n",
       "2                                        [IGHVF3-G12*04]   \n",
       "3                                        [IGHVF3-G12*04]   \n",
       "4                                        [IGHVF3-G12*04]   \n",
       "...                                                  ...   \n",
       "32665                                     [IGHVF3-G7*03]   \n",
       "32666                                     [IGHVF3-G7*03]   \n",
       "32667  [IGHVF3-G7*03, IGHVF3-G6*07, IGHVF3-G7*05, IGH...   \n",
       "32668                                     [IGHVF3-G7*03]   \n",
       "32669                                     [IGHVF3-G7*03]   \n",
       "\n",
       "                                                  d_call      j_call  \\\n",
       "0                             [IGHD2-21*01, IGHD2-21*02]  [IGHJ5*02]   \n",
       "1                             [IGHD2-21*01, IGHD2-21*02]  [IGHJ6*03]   \n",
       "2                                          [IGHD2-21*01]  [IGHJ1*01]   \n",
       "3      [IGHD2-21*01, IGHD2-15*01, IGHD2-8*02, IGHD2-2...  [IGHJ4*02]   \n",
       "4                                              [Short-D]  [IGHJ6*02]   \n",
       "...                                                  ...         ...   \n",
       "32665                                       [IGHD3-3*01]  [IGHJ3*02]   \n",
       "32666                                       [IGHD3-3*01]  [IGHJ6*03]   \n",
       "32667                                       [IGHD3-3*01]  [IGHJ6*03]   \n",
       "32668                                       [IGHD3-3*01]  [IGHJ5*02]   \n",
       "32669                                       [IGHD3-3*01]  [IGHJ5*02]   \n",
       "\n",
       "           c_call  v_sequence_start  v_sequence_end  d_sequence_start  \\\n",
       "0      [IGHA1*01]                 0             301               303   \n",
       "1      [IGHG4*06]                 0             301               303   \n",
       "2      [IGHG3*17]                 0             178               202   \n",
       "3      [IGHG3*23]                 0             299               311   \n",
       "4      [IGHG3*01]                 0             194               202   \n",
       "...           ...               ...             ...               ...   \n",
       "32665  [IGHG3*09]                 0             296               310   \n",
       "32666  [IGHG2*05]               143             441               451   \n",
       "32667   [IGHM*04]                 0             110               124   \n",
       "32668   [IGHE*03]                 0             296               299   \n",
       "32669  [IGHG3*13]                 0             296               306   \n",
       "\n",
       "       d_sequence_end  j_sequence_start  ...  c_trim_3  productive  \\\n",
       "0                 308               321  ...      35.0       False   \n",
       "1                 314               323  ...      37.0       False   \n",
       "2                 214               217  ...       NaN       False   \n",
       "3                 317               327  ...      20.0       False   \n",
       "4                 205               206  ...       1.0       False   \n",
       "...               ...               ...  ...       ...         ...   \n",
       "32665             320               320  ...      11.0       False   \n",
       "32666             477               486  ...      32.0       False   \n",
       "32667             144               147  ...      22.0       False   \n",
       "32668             314               315  ...       6.0       False   \n",
       "32669             334               340  ...       5.0       False   \n",
       "\n",
       "       stop_codon  vj_in_frame  \\\n",
       "0            True        False   \n",
       "1            True        False   \n",
       "2            True        False   \n",
       "3            True        False   \n",
       "4            True        False   \n",
       "...           ...          ...   \n",
       "32665       False        False   \n",
       "32666        True        False   \n",
       "32667        True        False   \n",
       "32668        True        False   \n",
       "32669        True        False   \n",
       "\n",
       "                                                    note  corruption_event  \\\n",
       "0                                    Stop codon present.     no-corruption   \n",
       "1                                    Stop codon present.     no-corruption   \n",
       "2      Stop codon present.Junction length not divisib...            remove   \n",
       "3                                                            no-corruption   \n",
       "4                                    Stop codon present.            remove   \n",
       "...                                                  ...               ...   \n",
       "32665                            V second C not present.     no-corruption   \n",
       "32666  Stop codon present.Junction length not divisib...               add   \n",
       "32667  Stop codon present.Junction length not divisib...            remove   \n",
       "32668                                                        no-corruption   \n",
       "32669  Stop codon present.Junction length not divisib...     no-corruption   \n",
       "\n",
       "       corruption_add_amount  corruption_remove_amount  \\\n",
       "0                          0                         0   \n",
       "1                          0                         0   \n",
       "2                          0                       121   \n",
       "3                          0                         0   \n",
       "4                          0                       105   \n",
       "...                      ...                       ...   \n",
       "32665                      0                         0   \n",
       "32666                    143                         0   \n",
       "32667                      0                       188   \n",
       "32668                      0                         0   \n",
       "32669                      0                         0   \n",
       "\n",
       "                              corruption_removed_section  \\\n",
       "0                                                          \n",
       "1                                                          \n",
       "2      AAGCCGCAGAAGCAGTCCTCCCGCTAAGGACTCACGCAGTCTTTAC...   \n",
       "3                                                          \n",
       "4      GATCTGCAGCTGCAGGAGTTCGGCTCAAGCATTCTGGTGCCTACAC...   \n",
       "...                                                  ...   \n",
       "32665                                                      \n",
       "32666                                                      \n",
       "32667  CAGGTTCAGCCGTAGGAGTCGGGCACAAGACTCGTGAAGCCTTCTG...   \n",
       "32668                                                      \n",
       "32669                                                      \n",
       "\n",
       "                                corruption_added_section  \n",
       "0                                                         \n",
       "1                                                         \n",
       "2                                                         \n",
       "3                                                         \n",
       "4                                                         \n",
       "...                                                  ...  \n",
       "32665                                                     \n",
       "32666  CTACGGATTTTCACCGTCACCCACGTTTCAAAATGCTCCGTCGCAT...  \n",
       "32667                                                     \n",
       "32668                                                     \n",
       "32669                                                     \n",
       "\n",
       "[32670 rows x 39 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>v_call</th>\n",
       "      <th>d_call</th>\n",
       "      <th>j_call</th>\n",
       "      <th>c_call</th>\n",
       "      <th>v_sequence_start</th>\n",
       "      <th>v_sequence_end</th>\n",
       "      <th>d_sequence_start</th>\n",
       "      <th>d_sequence_end</th>\n",
       "      <th>j_sequence_start</th>\n",
       "      <th>...</th>\n",
       "      <th>c_trim_3</th>\n",
       "      <th>productive</th>\n",
       "      <th>stop_codon</th>\n",
       "      <th>vj_in_frame</th>\n",
       "      <th>note</th>\n",
       "      <th>corruption_event</th>\n",
       "      <th>corruption_add_amount</th>\n",
       "      <th>corruption_remove_amount</th>\n",
       "      <th>corruption_removed_section</th>\n",
       "      <th>corruption_added_section</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAGATGCAGCTCCAGGAGTCAAGCTCACGGCTGGCCTGGCCTTTAC...</td>\n",
       "      <td>[IGHVF3-G12*04]</td>\n",
       "      <td>[IGHD2-21*01, IGHD2-21*02]</td>\n",
       "      <td>[IGHJ5*02]</td>\n",
       "      <td>[IGHA1*01]</td>\n",
       "      <td>0</td>\n",
       "      <td>301</td>\n",
       "      <td>303</td>\n",
       "      <td>308</td>\n",
       "      <td>321</td>\n",
       "      <td>...</td>\n",
       "      <td>35.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Stop codon present.</td>\n",
       "      <td>no-corruption</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CAGCTGCAGCTGCAGAGAGTCCGGCTCAGGACTGGTGAAGCCTTCA...</td>\n",
       "      <td>[IGHVF3-G12*04]</td>\n",
       "      <td>[IGHD2-21*01, IGHD2-21*02]</td>\n",
       "      <td>[IGHJ6*03]</td>\n",
       "      <td>[IGHG4*06]</td>\n",
       "      <td>0</td>\n",
       "      <td>301</td>\n",
       "      <td>303</td>\n",
       "      <td>314</td>\n",
       "      <td>323</td>\n",
       "      <td>...</td>\n",
       "      <td>37.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Stop codon present.</td>\n",
       "      <td>no-corruption</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AGTCATCGGTAAAGGCCCTTGAGGGGAGAGGGTGCATCAATCCTAA...</td>\n",
       "      <td>[IGHVF3-G12*04]</td>\n",
       "      <td>[IGHD2-21*01]</td>\n",
       "      <td>[IGHJ1*01]</td>\n",
       "      <td>[IGHG3*17]</td>\n",
       "      <td>0</td>\n",
       "      <td>178</td>\n",
       "      <td>202</td>\n",
       "      <td>214</td>\n",
       "      <td>217</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Stop codon present.Junction length not divisib...</td>\n",
       "      <td>remove</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>AAGCCGCAGAAGCAGTCCTCCCGCTAAGGACTCACGCAGTCTTTAC...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CAGCTGCAGCTGCGTGAGTCCGGCTCAGGACTGGTGAAGCCTTCAC...</td>\n",
       "      <td>[IGHVF3-G12*04]</td>\n",
       "      <td>[IGHD2-21*01, IGHD2-15*01, IGHD2-8*02, IGHD2-2...</td>\n",
       "      <td>[IGHJ4*02]</td>\n",
       "      <td>[IGHG3*23]</td>\n",
       "      <td>0</td>\n",
       "      <td>299</td>\n",
       "      <td>311</td>\n",
       "      <td>317</td>\n",
       "      <td>327</td>\n",
       "      <td>...</td>\n",
       "      <td>20.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>no-corruption</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TGGAGCTGGATCCGGAGGTCAGAAGGGAAGGGCCTGGAGAGGAGAG...</td>\n",
       "      <td>[IGHVF3-G12*04]</td>\n",
       "      <td>[Short-D]</td>\n",
       "      <td>[IGHJ6*02]</td>\n",
       "      <td>[IGHG3*01]</td>\n",
       "      <td>0</td>\n",
       "      <td>194</td>\n",
       "      <td>202</td>\n",
       "      <td>205</td>\n",
       "      <td>206</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Stop codon present.</td>\n",
       "      <td>remove</td>\n",
       "      <td>0</td>\n",
       "      <td>105</td>\n",
       "      <td>GATCTGCAGCTGCAGGAGTTCGGCTCAAGCATTCTGGTGCCTACAC...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32665</th>\n",
       "      <td>CATGTGCAGCTGCATGAGTCGGGCCCTGAACTGTCTAGGCCTTCGC...</td>\n",
       "      <td>[IGHVF3-G7*03]</td>\n",
       "      <td>[IGHD3-3*01]</td>\n",
       "      <td>[IGHJ3*02]</td>\n",
       "      <td>[IGHG3*09]</td>\n",
       "      <td>0</td>\n",
       "      <td>296</td>\n",
       "      <td>310</td>\n",
       "      <td>320</td>\n",
       "      <td>320</td>\n",
       "      <td>...</td>\n",
       "      <td>11.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>V second C not present.</td>\n",
       "      <td>no-corruption</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32666</th>\n",
       "      <td>CTACGGATTTTCACCGTCACCCACGTTTCAAAATGCTCCGTCGCAT...</td>\n",
       "      <td>[IGHVF3-G7*03]</td>\n",
       "      <td>[IGHD3-3*01]</td>\n",
       "      <td>[IGHJ6*03]</td>\n",
       "      <td>[IGHG2*05]</td>\n",
       "      <td>143</td>\n",
       "      <td>441</td>\n",
       "      <td>451</td>\n",
       "      <td>477</td>\n",
       "      <td>486</td>\n",
       "      <td>...</td>\n",
       "      <td>32.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Stop codon present.Junction length not divisib...</td>\n",
       "      <td>add</td>\n",
       "      <td>143</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>CTACGGATTTTCACCGTCACCCACGTTTCAAAATGCTCCGTCGCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32667</th>\n",
       "      <td>CTCCCTCAAGAGTCGAGTCATCATATCAGTTGACCCGTCTAAGAAC...</td>\n",
       "      <td>[IGHVF3-G7*03, IGHVF3-G6*07, IGHVF3-G7*05, IGH...</td>\n",
       "      <td>[IGHD3-3*01]</td>\n",
       "      <td>[IGHJ6*03]</td>\n",
       "      <td>[IGHM*04]</td>\n",
       "      <td>0</td>\n",
       "      <td>110</td>\n",
       "      <td>124</td>\n",
       "      <td>144</td>\n",
       "      <td>147</td>\n",
       "      <td>...</td>\n",
       "      <td>22.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Stop codon present.Junction length not divisib...</td>\n",
       "      <td>remove</td>\n",
       "      <td>0</td>\n",
       "      <td>188</td>\n",
       "      <td>CAGGTTCAGCCGTAGGAGTCGGGCACAAGACTCGTGAAGCCTTCTG...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32668</th>\n",
       "      <td>CAGGTGCAGCTTCAGGAGNCGGGCCCAGGACCGGTGATGCGCTCTG...</td>\n",
       "      <td>[IGHVF3-G7*03]</td>\n",
       "      <td>[IGHD3-3*01]</td>\n",
       "      <td>[IGHJ5*02]</td>\n",
       "      <td>[IGHE*03]</td>\n",
       "      <td>0</td>\n",
       "      <td>296</td>\n",
       "      <td>299</td>\n",
       "      <td>314</td>\n",
       "      <td>315</td>\n",
       "      <td>...</td>\n",
       "      <td>6.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>no-corruption</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32669</th>\n",
       "      <td>CAGNTGAAGGTGCAGTAGTCGGGGCCGGACTGGTGAAGCCTTGGGA...</td>\n",
       "      <td>[IGHVF3-G7*03]</td>\n",
       "      <td>[IGHD3-3*01]</td>\n",
       "      <td>[IGHJ5*02]</td>\n",
       "      <td>[IGHG3*13]</td>\n",
       "      <td>0</td>\n",
       "      <td>296</td>\n",
       "      <td>306</td>\n",
       "      <td>334</td>\n",
       "      <td>340</td>\n",
       "      <td>...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Stop codon present.Junction length not divisib...</td>\n",
       "      <td>no-corruption</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32670 rows × 39 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 9
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
