<!DOCTYPE html>
<html class="writer-html5" lang="python" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>GenAIRR.utilities.asc_utilities &mdash; GenAIRR  documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=92fd9be5" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../../_static/documentation_options.js?v=d048f138"></script>
        <script src="../../../_static/doctools.js?v=888ff710"></script>
        <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            GenAIRR
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html">src</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">GenAIRR</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">GenAIRR.utilities.asc_utilities</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for GenAIRR.utilities.asc_utilities</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>
<span class="kn">from</span> <span class="nn">..utilities.data_utilities</span> <span class="kn">import</span> <span class="n">create_allele_dict</span>
<span class="kn">from</span> <span class="nn">scipy.cluster.hierarchy</span> <span class="kn">import</span> <span class="n">dendrogram</span><span class="p">,</span> <span class="n">linkage</span><span class="p">,</span> <span class="n">fcluster</span>
<span class="kn">from</span> <span class="nn">scipy.spatial.distance</span> <span class="kn">import</span> <span class="n">squareform</span>
<span class="kn">from</span> <span class="nn">..alleles.allele</span> <span class="kn">import</span> <span class="n">VAllele</span>

<div class="viewcode-block" id="allele_diff">
<a class="viewcode-back" href="../../../GenAIRR.utilities.html#GenAIRR.utilities.asc_utilities.allele_diff">[docs]</a>
<span class="k">def</span> <span class="nf">allele_diff</span><span class="p">(</span><span class="n">reference_allele</span><span class="p">,</span> <span class="n">sample_allele</span><span class="p">,</span> <span class="n">position_threshold</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">snps</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Identifies differences between two alleles, optionally considering only positions beyond a threshold and reporting single nucleotide polymorphisms (SNPs).</span>

<span class="sd">        Args:</span>
<span class="sd">            reference_allele (str): The sequence of the reference allele.</span>
<span class="sd">            sample_allele (str): The sequence of the sample allele to compare against the reference.</span>
<span class="sd">            position_threshold (int, optional): The position from which to start comparing the sequences. Defaults to 0, which means comparison starts from the beginning.</span>
<span class="sd">            snps (bool, optional): If True, differences are reported as SNPs in the format &#39;RefPositionSample&#39;. If False, only the positions of differences are reported. Defaults to True.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: A list of SNPs or position indices where the two alleles differ, considering the specified threshold and SNP reporting preference.</span>
<span class="sd">        &quot;&quot;&quot;</span>
    <span class="n">germs</span> <span class="o">=</span> <span class="p">[</span><span class="n">reference_allele</span><span class="p">,</span> <span class="n">sample_allele</span><span class="p">]</span>
    <span class="n">max_length</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">germ</span><span class="p">)</span> <span class="k">for</span> <span class="n">germ</span> <span class="ow">in</span> <span class="n">germs</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">germs</span><span class="p">)):</span>
        <span class="n">germs</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+=</span> <span class="s1">&#39;.&#39;</span> <span class="o">*</span> <span class="p">(</span><span class="n">max_length</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">germs</span><span class="p">[</span><span class="n">i</span><span class="p">]))</span>
    <span class="k">def</span> <span class="nf">setdiff_mat</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
        <span class="n">unique_chars</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
        <span class="n">filter_chars</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="s1">&#39;N&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">}</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="n">unique_chars</span> <span class="o">-</span> <span class="n">filter_chars</span><span class="p">)</span>
    <span class="n">idx_strings</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">max_length</span><span class="p">):</span>
        <span class="n">column_chars</span> <span class="o">=</span> <span class="p">[</span><span class="n">germ</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">germ</span> <span class="ow">in</span> <span class="n">germs</span><span class="p">]</span>
        <span class="n">diff_count</span> <span class="o">=</span> <span class="n">setdiff_mat</span><span class="p">(</span><span class="n">column_chars</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">diff_count</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">i</span> <span class="o">&gt;=</span> <span class="p">(</span><span class="n">position_threshold</span><span class="o">-</span><span class="mi">1</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">snps</span><span class="p">:</span>
                <span class="n">concatenated_str</span> <span class="o">=</span> <span class="n">column_chars</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span> <span class="o">+</span> <span class="n">column_chars</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
                <span class="n">idx_strings</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">concatenated_str</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">idx_strings</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">idx_strings</span></div>


<div class="viewcode-block" id="hamming_distance">
<a class="viewcode-back" href="../../../GenAIRR.utilities.html#GenAIRR.utilities.asc_utilities.hamming_distance">[docs]</a>
<span class="k">def</span> <span class="nf">hamming_distance</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span> <span class="n">s2</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculates the Hamming distance between two sequences, ignoring positions with &#39;N&#39; and gaps represented by &#39;.&#39;.</span>

<span class="sd">        Args:</span>
<span class="sd">            s1 (str): The first sequence for comparison.</span>
<span class="sd">            s2 (str): The second sequence for comparison.</span>

<span class="sd">        Returns:</span>
<span class="sd">            float: The Hamming distance between the two sequences, normalized by the number of comparable positions (excluding gaps).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">s1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">s1</span><span class="p">))</span>
    <span class="n">s2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">s2</span><span class="p">))</span>
    <span class="n">mismatches</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">s1</span> <span class="o">!=</span> <span class="s1">&#39;N&#39;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">s2</span> <span class="o">!=</span> <span class="s1">&#39;N&#39;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">s1</span> <span class="o">!=</span> <span class="n">s2</span><span class="p">))</span>
    <span class="n">gapGapMatches</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">s1</span> <span class="o">==</span> <span class="s1">&#39;.&#39;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">s2</span> <span class="o">==</span> <span class="s1">&#39;.&#39;</span><span class="p">))</span>
    <span class="n">count</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">s1</span><span class="p">)</span> <span class="o">-</span> <span class="n">gapGapMatches</span>    
    <span class="n">distance</span> <span class="o">=</span> <span class="n">mismatches</span> <span class="o">/</span> <span class="n">count</span> <span class="k">if</span> <span class="n">count</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>
    <span class="k">return</span> <span class="n">distance</span></div>


<span class="sd">&quot;&quot;&quot; def asc_distance(germline_set):</span>
<span class="sd">    # Check sequences length. If not even pad the sequences to the max length</span>
<span class="sd">    max_length = max(len(s) for s in germline_set.values())</span>
<span class="sd">    germline_set_padded = {allele: sequence.ljust(max_length, &#39;N&#39;) for allele, sequence in germline_set.items()}</span>
<span class="sd">    # Compute the distance between pairs. penalize for gaps</span>
<span class="sd">    germline_distance = np.zeros((len(germline_set_padded), len(germline_set_padded)))</span>
<span class="sd">    sequences = list(germline_set_padded.values())</span>
<span class="sd">    for i in range(len(sequences)):</span>
<span class="sd">        s1 = np.array(list(sequences[i]))</span>
<span class="sd">        germline_distance[i, :] = np.array([hamming_distance(s1, np.array(list(s2))) for s2 in sequences])</span>
<span class="sd">    return germline_distance &quot;&quot;&quot;</span>

<div class="viewcode-block" id="asc_distance">
<a class="viewcode-back" href="../../../GenAIRR.utilities.html#GenAIRR.utilities.asc_utilities.asc_distance">[docs]</a>
<span class="k">def</span> <span class="nf">asc_distance</span><span class="p">(</span><span class="n">germline_set</span><span class="p">,</span> <span class="n">trim_3prime_side</span><span class="o">=</span><span class="mi">318</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculates the pairwise Hamming distances between alleles in a set, optionally trimming sequences to a specified length before comparison.</span>

<span class="sd">        Args:</span>
<span class="sd">            germline_set (dict): A dictionary of allele sequences with allele names as keys.</span>
<span class="sd">            trim_3prime_side (int, optional): The position up to which sequences should be considered. If None, the entire length is used. Defaults to 318.</span>

<span class="sd">        Returns:</span>
<span class="sd">            ndarray: A square matrix of Hamming distances between all pairs of alleles in the set.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">trim_3prime_side</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">germline_set</span> <span class="o">=</span> <span class="p">{</span><span class="n">allele</span><span class="p">:</span> <span class="n">seq</span><span class="p">[:</span><span class="n">trim_3prime_side</span><span class="p">]</span> <span class="k">for</span> <span class="n">allele</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">germline_set</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
    <span class="n">max_length</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">germline_set</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
    <span class="n">germline_set_padded</span> <span class="o">=</span> <span class="p">{</span><span class="n">allele</span><span class="p">:</span> <span class="n">sequence</span><span class="o">.</span><span class="n">ljust</span><span class="p">(</span><span class="n">max_length</span><span class="p">,</span> <span class="s1">&#39;N&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">allele</span><span class="p">,</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">germline_set</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
    
    <span class="n">sequences_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="nb">list</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">germline_set_padded</span><span class="o">.</span><span class="n">values</span><span class="p">()])</span>

    <span class="n">num_sequences</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequences_array</span><span class="p">)</span>
    <span class="n">germline_distance_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">num_sequences</span><span class="p">,</span> <span class="n">num_sequences</span><span class="p">))</span>
    
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_sequences</span><span class="p">):</span>
        <span class="n">s1</span> <span class="o">=</span> <span class="n">sequences_array</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_sequences</span><span class="p">):</span>
            <span class="n">s2</span> <span class="o">=</span> <span class="n">sequences_array</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
            <span class="n">mismatches</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">s1</span> <span class="o">!=</span> <span class="s1">&#39;N&#39;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">s2</span> <span class="o">!=</span> <span class="s1">&#39;N&#39;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">s1</span> <span class="o">!=</span> <span class="n">s2</span><span class="p">))</span>
            <span class="n">gapGapMatches</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">s1</span> <span class="o">==</span> <span class="s1">&#39;.&#39;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">s2</span> <span class="o">==</span> <span class="s1">&#39;.&#39;</span><span class="p">))</span>
            <span class="n">count</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">s1</span><span class="p">)</span> <span class="o">-</span> <span class="n">gapGapMatches</span>
            <span class="n">germline_distance_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">mismatches</span> <span class="o">/</span> <span class="n">count</span> <span class="k">if</span> <span class="n">count</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>

    <span class="k">return</span> <span class="n">germline_distance_matrix</span></div>


<div class="viewcode-block" id="asc_clust">
<a class="viewcode-back" href="../../../GenAIRR.utilities.html#GenAIRR.utilities.asc_utilities.asc_clust">[docs]</a>
<span class="k">def</span> <span class="nf">asc_clust</span><span class="p">(</span><span class="n">germline_distance</span><span class="p">,</span> <span class="n">germline_set</span><span class="p">,</span> <span class="n">family_threshold</span> <span class="o">=</span> <span class="mi">75</span><span class="p">,</span> <span class="n">allele_cluster_threshold</span> <span class="o">=</span> <span class="mi">95</span><span class="p">,</span> <span class="n">trim_3prime_side</span> <span class="o">=</span> <span class="mi">318</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Performs hierarchical clustering on alleles based on their Hamming distances and classifies them into families and allele clusters.</span>

<span class="sd">        Args:</span>
<span class="sd">            germline_distance (ndarray): A square matrix of pairwise Hamming distances between alleles.</span>
<span class="sd">            germline_set (dict): A dictionary of allele sequences with allele names as keys.</span>
<span class="sd">            family_threshold (int, optional): The distance threshold for classifying alleles into families. Defaults to 75.</span>
<span class="sd">            allele_cluster_threshold (int, optional): The distance threshold for classifying alleles within families into clusters. Defaults to 95.</span>
<span class="sd">            trim_3prime_side (int, optional): The position up to which sequences should be considered for clustering. Defaults to 318.</span>

<span class="sd">        Returns:</span>
<span class="sd">            DataFrame: A pandas DataFrame with columns for family and allele cluster assignments, allele names, and additional columns for duplicated alleles and differences past the trim position.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">condensed_distance</span> <span class="o">=</span> <span class="n">squareform</span><span class="p">(</span><span class="n">germline_distance</span><span class="p">)</span>
    <span class="n">germline_cluster</span> <span class="o">=</span> <span class="n">linkage</span><span class="p">(</span><span class="n">condensed_distance</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;complete&#39;</span><span class="p">)</span>
    <span class="n">dendro</span> <span class="o">=</span> <span class="n">dendrogram</span><span class="p">(</span><span class="n">germline_cluster</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">germline_set</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="n">no_plot</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">labels_order</span> <span class="o">=</span> <span class="n">dendro</span><span class="p">[</span><span class="s1">&#39;ivl&#39;</span><span class="p">]</span>
    <span class="n">leaf_order</span> <span class="o">=</span> <span class="n">dendro</span><span class="p">[</span><span class="s1">&#39;leaves&#39;</span><span class="p">]</span>
    
    <span class="n">labels</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">germline_set</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
    <span class="n">segment</span> <span class="o">=</span> <span class="n">labels</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="mi">4</span><span class="p">]</span>
    
    <span class="n">thresh</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="o">-</span><span class="n">family_threshold</span><span class="o">/</span><span class="mi">100</span><span class="p">,</span><span class="mi">1</span><span class="o">-</span><span class="n">allele_cluster_threshold</span><span class="o">/</span><span class="mi">100</span><span class="p">]</span>
    <span class="n">family_cluster</span> <span class="o">=</span> <span class="n">fcluster</span><span class="p">(</span><span class="n">germline_cluster</span><span class="p">,</span> 
                        <span class="n">t</span><span class="o">=</span><span class="n">thresh</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> 
                        <span class="n">criterion</span><span class="o">=</span><span class="s1">&#39;distance&#39;</span><span class="p">)</span>
    <span class="n">family_clusters_in_order</span> <span class="o">=</span> <span class="n">family_cluster</span><span class="p">[</span><span class="n">leaf_order</span><span class="p">]</span>
    <span class="n">family_clusters_in_order_renumbered</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">factorize</span><span class="p">(</span><span class="n">family_clusters_in_order</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span>
    
    <span class="n">allele_cluster</span> <span class="o">=</span> <span class="n">fcluster</span><span class="p">(</span><span class="n">germline_cluster</span><span class="p">,</span> 
                            <span class="n">t</span><span class="o">=</span><span class="n">thresh</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> 
                            <span class="n">criterion</span><span class="o">=</span><span class="s1">&#39;distance&#39;</span><span class="p">)</span>
    <span class="n">allele_clusters_in_order</span> <span class="o">=</span> <span class="n">allele_cluster</span><span class="p">[</span><span class="n">leaf_order</span><span class="p">]</span>
    <span class="n">allele_clusters_in_order_renumbered</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">factorize</span><span class="p">(</span><span class="n">allele_clusters_in_order</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span>
    
    <span class="n">alleleClusterTable</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
                <span class="s1">&#39;Family&#39;</span><span class="p">:</span> <span class="n">family_clusters_in_order_renumbered</span><span class="p">,</span>
                <span class="s1">&#39;Allele_Cluster&#39;</span><span class="p">:</span> <span class="n">allele_clusters_in_order_renumbered</span><span class="p">,</span>
                <span class="s1">&#39;Allele&#39;</span><span class="p">:</span> <span class="n">labels_order</span>
            <span class="p">})</span>
    
    <span class="n">alleleClusterTable</span><span class="p">[</span><span class="s1">&#39;duplicated_allele&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">NA</span>
    <span class="n">alleleClusterTable</span><span class="p">[</span><span class="s1">&#39;diff_pos_past_trim&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">NA</span>
    
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">germline_cluster</span><span class="p">):</span>
        <span class="n">cluster_1</span><span class="p">,</span> <span class="n">cluster_2</span><span class="p">,</span> <span class="n">distance</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">row</span>
        <span class="k">if</span> <span class="n">distance</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># check which allele is longer, this should be the reference allele</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">germline_set</span><span class="p">[</span><span class="n">labels</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">cluster_1</span><span class="p">)]])</span> <span class="o">&gt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">germline_set</span><span class="p">[</span><span class="n">labels</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">cluster_2</span><span class="p">)]]):</span>
                <span class="n">cluster_1_label</span> <span class="o">=</span> <span class="n">labels</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">cluster_1</span><span class="p">)]</span>
                <span class="n">cluster_2_label</span> <span class="o">=</span> <span class="n">labels</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">cluster_2</span><span class="p">)]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">cluster_2_label</span> <span class="o">=</span> <span class="n">labels</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">cluster_1</span><span class="p">)]</span>
                <span class="n">cluster_1_label</span> <span class="o">=</span> <span class="n">labels</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">cluster_2</span><span class="p">)]</span>
            <span class="c1"># check if the &#39;duplicated allele&#39; is seperated passed the 3prime trim</span>
            <span class="k">if</span> <span class="n">trim_3prime_side</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">snps</span> <span class="o">=</span> <span class="n">allele_diff</span><span class="p">(</span><span class="n">germline_set</span><span class="p">[</span><span class="n">cluster_1_label</span><span class="p">],</span> <span class="n">germline_set</span><span class="p">[</span><span class="n">cluster_2_label</span><span class="p">],</span> <span class="n">trim_3prime_side</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">snps</span><span class="p">:</span>
                    <span class="n">alleleClusterTable</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">alleleClusterTable</span><span class="p">[</span><span class="s1">&#39;Allele&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">cluster_2_label</span><span class="p">,</span> <span class="s1">&#39;diff_pos_past_trim&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="n">snps</span><span class="p">))</span>
                    <span class="n">alleleClusterTable</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">alleleClusterTable</span><span class="p">[</span><span class="s1">&#39;Allele&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">cluster_1_label</span><span class="p">,</span> <span class="s1">&#39;duplicated_allele&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cluster_1_label</span><span class="o">+</span><span class="s2">&quot;,&quot;</span><span class="o">+</span><span class="n">cluster_2_label</span>
                    <span class="n">alleleClusterTable</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">alleleClusterTable</span><span class="p">[</span><span class="s1">&#39;Allele&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">cluster_2_label</span><span class="p">,</span> <span class="s1">&#39;duplicated_allele&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cluster_1_label</span><span class="o">+</span><span class="s2">&quot;,&quot;</span><span class="o">+</span><span class="n">cluster_2_label</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">duplicate_cell_is_na</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">isna</span><span class="p">(</span><span class="n">alleleClusterTable</span><span class="p">[</span><span class="s1">&#39;duplicated_allele&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">alleleClusterTable</span><span class="p">[</span><span class="s1">&#39;Allele&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">cluster_1_label</span><span class="p">])</span><span class="o">.</span><span class="n">any</span><span class="p">()</span>
                    <span class="c1"># check if the duplicated allele is already set</span>
                    <span class="k">if</span> <span class="o">~</span><span class="n">duplicate_cell_is_na</span><span class="p">:</span>
                        <span class="n">alleleClusterTable</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">alleleClusterTable</span><span class="p">[</span><span class="s1">&#39;Allele&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">cluster_1_label</span><span class="p">,</span> <span class="s1">&#39;duplicated_allele&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="n">cluster_2_label</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">alleleClusterTable</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">alleleClusterTable</span><span class="p">[</span><span class="s1">&#39;Allele&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">cluster_1_label</span><span class="p">,</span> <span class="s1">&#39;duplicated_allele&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cluster_2_label</span>
                        <span class="n">alleleClusterTable</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">alleleClusterTable</span><span class="p">[</span><span class="s1">&#39;Allele&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">cluster_2_label</span><span class="p">,</span> <span class="s1">&#39;duplicated_allele&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;remove&#39;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">duplicate_cell_is_na</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">isna</span><span class="p">(</span><span class="n">alleleClusterTable</span><span class="p">[</span><span class="s1">&#39;duplicated_allele&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">alleleClusterTable</span><span class="p">[</span><span class="s1">&#39;Allele&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">cluster_1_label</span><span class="p">])</span><span class="o">.</span><span class="n">any</span><span class="p">()</span>
                <span class="c1"># check if the duplicated allele is already set</span>
                <span class="k">if</span> <span class="o">~</span><span class="n">duplicate_cell_is_na</span><span class="p">:</span>
                    <span class="n">alleleClusterTable</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">alleleClusterTable</span><span class="p">[</span><span class="s1">&#39;Allele&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">cluster_1_label</span><span class="p">,</span> <span class="s1">&#39;duplicated_allele&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="n">cluster_2_label</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">alleleClusterTable</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">alleleClusterTable</span><span class="p">[</span><span class="s1">&#39;Allele&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">cluster_1_label</span><span class="p">,</span> <span class="s1">&#39;duplicated_allele&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cluster_2_label</span>
                    <span class="n">alleleClusterTable</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">alleleClusterTable</span><span class="p">[</span><span class="s1">&#39;Allele&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">cluster_2_label</span><span class="p">,</span> <span class="s1">&#39;duplicated_allele&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;remove&#39;</span>
    
    <span class="c1"># rename the alleles (segment-FamilyCluster-AlleleCluster*allele)</span>
    <span class="n">alleleClusterTable_copy</span> <span class="o">=</span> <span class="n">alleleClusterTable</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">alleleClusterTable_copy</span> <span class="o">=</span> <span class="n">alleleClusterTable_copy</span><span class="p">[</span><span class="n">alleleClusterTable_copy</span><span class="p">[</span><span class="s1">&#39;duplicated_allele&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;remove&#39;</span><span class="p">]</span>
    <span class="n">alleleClusterTable_copy</span><span class="p">[</span><span class="s1">&#39;allele_idx&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">alleleClusterTable_copy</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;Allele_Cluster&#39;</span><span class="p">)[</span><span class="s1">&#39;Family&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>
    <span class="n">alleleClusterTable_copy</span><span class="p">[</span><span class="s1">&#39;allele_index&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">alleleClusterTable_copy</span><span class="p">[</span><span class="s1">&#39;allele_idx&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="c1"># Change &#39;allele_index&#39; value to the minimum for rows with the same &#39;duplicated_allele&#39;</span>
    <span class="n">alleleClusterTable_copy</span><span class="p">[</span><span class="s1">&#39;allele_index&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">alleleClusterTable_copy</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;Allele_Cluster&#39;</span><span class="p">,</span> <span class="s1">&#39;Family&#39;</span><span class="p">,</span> <span class="s1">&#39;duplicated_allele&#39;</span><span class="p">])</span>
        <span class="p">[</span><span class="s1">&#39;allele_idx&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="s1">&#39;min&#39;</span><span class="p">)</span>
    <span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="n">alleleClusterTable_copy</span><span class="p">[</span><span class="s1">&#39;allele_idx&#39;</span><span class="p">])</span>
    <span class="n">alleleClusterTable_copy</span><span class="p">[</span><span class="s1">&#39;allele_index&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">alleleClusterTable_copy</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;Allele_Cluster&#39;</span><span class="p">)[</span><span class="s1">&#39;allele_index&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">diff</span><span class="p">()</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">cumsum</span><span class="p">())</span>
    <span class="n">alleleClusterTable_copy</span> <span class="o">=</span> <span class="n">alleleClusterTable_copy</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s1">&#39;allele_idx&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">alleleClusterTable_copy</span><span class="p">[</span><span class="s1">&#39;new_allele&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">alleleClusterTable_copy</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">segment</span><span class="si">}</span><span class="s2">F</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Family&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">-G</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Allele_Cluster&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">*</span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;allele_index&#39;</span><span class="p">]))</span><span class="o">.</span><span class="n">zfill</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;allele_index&#39;</span><span class="p">]</span><span class="w"> </span><span class="o">&gt;=</span><span class="w"> </span><span class="mi">1</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="nb">str</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;allele_index&#39;</span><span class="p">])</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">alleleClusterTable_copy</span><span class="p">[</span><span class="s1">&#39;new_allele&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="n">alleleClusterTable_copy</span><span class="p">[</span><span class="s1">&#39;diff_pos_past_trim&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;_</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">notna</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">alleleClusterTable_copy</span></div>


<div class="viewcode-block" id="asc_dict">
<a class="viewcode-back" href="../../../GenAIRR.utilities.html#GenAIRR.utilities.asc_utilities.asc_dict">[docs]</a>
<span class="k">def</span> <span class="nf">asc_dict</span><span class="p">(</span><span class="n">allele_cluster_table</span><span class="p">,</span> <span class="n">germline_set</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Converts an allele cluster table into a dictionary of VAllele objects organized by allele cluster sequence (ASC).</span>

<span class="sd">        Args:</span>
<span class="sd">            allele_cluster_table (DataFrame): A pandas DataFrame containing allele family and cluster assignments.</span>
<span class="sd">            germline_set (dict): A dictionary of original allele sequences with allele names as keys.</span>

<span class="sd">        Returns:</span>
<span class="sd">            defaultdict: A dictionary where keys are ASC identifiers and values are lists of VAllele objects belonging to each ASC.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">germline_set_asc</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
        
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">allele_cluster_table</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="n">allele</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Allele&#39;</span><span class="p">]</span>
            <span class="n">new_allele</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;new_allele&#39;</span><span class="p">]</span>
            <span class="c1"># get asc</span>
            <span class="n">asc</span> <span class="o">=</span> <span class="n">new_allele</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;*&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">seq</span> <span class="o">=</span> <span class="n">germline_set</span><span class="p">[</span><span class="n">allele</span><span class="p">]</span>
            <span class="n">ungapped_length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">))</span>
            <span class="n">germline_set_asc</span><span class="p">[</span><span class="n">asc</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">VAllele</span><span class="p">(</span><span class="n">new_allele</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="n">ungapped_length</span><span class="p">))</span>
                
        <span class="k">return</span> <span class="n">germline_set_asc</span></div>

    
<div class="viewcode-block" id="create_asc_germline_set">
<a class="viewcode-back" href="../../../GenAIRR.utilities.html#GenAIRR.utilities.asc_utilities.create_asc_germline_set">[docs]</a>
<span class="k">def</span> <span class="nf">create_asc_germline_set</span><span class="p">(</span><span class="n">user_reference</span><span class="p">,</span> <span class="n">segment</span> <span class="o">=</span> <span class="s2">&quot;V&quot;</span><span class="p">,</span> <span class="n">trim_3prime_side</span> <span class="o">=</span> <span class="mi">318</span><span class="p">,</span> <span class="n">family_threshold</span> <span class="o">=</span> <span class="mi">75</span><span class="p">,</span> <span class="n">allele_cluster_threshold</span> <span class="o">=</span> <span class="mi">95</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates an ASC (Allele Sequence Cluster) germline set from a user-provided reference, applying trimming and clustering thresholds.</span>

<span class="sd">        Args:</span>
<span class="sd">            user_reference (dict or path): A dictionary of allele sequences or a path to a fasta file containing alleles.</span>
<span class="sd">            segment (str, optional): The segment type to consider (e.g., &quot;V&quot;). Defaults to &quot;V&quot;.</span>
<span class="sd">            trim_3prime_side (int, optional): The position up to which sequences should be considered for clustering. Defaults to 318.</span>
<span class="sd">            family_threshold (int, optional): The distance threshold for classifying alleles into families. Defaults to 75.</span>
<span class="sd">            allele_cluster_threshold (int, optional): The distance threshold for classifying alleles within families into clusters. Defaults to 95.</span>

<span class="sd">        Returns:</span>
<span class="sd">            tuple: A tuple containing the ASC germline set dictionary and the allele cluster table DataFrame.</span>
<span class="sd">        &quot;&quot;&quot;</span>
    <span class="n">germline_set_user</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">dict</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">create_allele_dict</span><span class="p">(</span><span class="n">user_reference</span><span class="p">)</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">val</span> <span class="ow">in</span> <span class="n">value</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">segment</span> <span class="ow">in</span> <span class="n">val</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
                <span class="n">germline_set_user</span><span class="p">[</span><span class="n">val</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">val</span><span class="o">.</span><span class="n">gapped_seq</span>
    
    <span class="n">germline_distance</span> <span class="o">=</span> <span class="n">asc_distance</span><span class="p">(</span><span class="n">germline_set_user</span><span class="p">)</span>

    <span class="n">asc_table</span> <span class="o">=</span> <span class="n">asc_clust</span><span class="p">(</span><span class="n">germline_distance</span><span class="p">,</span> <span class="n">germline_set_user</span><span class="p">,</span> <span class="n">family_threshold</span><span class="p">,</span> <span class="n">allele_cluster_threshold</span><span class="p">,</span> <span class="n">trim_3prime_side</span><span class="p">)</span>

    <span class="n">germline_set_asc</span> <span class="o">=</span> <span class="n">asc_dict</span><span class="p">(</span><span class="n">asc_table</span><span class="p">,</span> <span class="n">germline_set_user</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">germline_set_asc</span><span class="p">,</span><span class="n">asc_table</span></div>

    
<div class="viewcode-block" id="test_asc">
<a class="viewcode-back" href="../../../GenAIRR.utilities.html#GenAIRR.utilities.asc_utilities.test_asc">[docs]</a>
<span class="k">def</span> <span class="nf">test_asc</span><span class="p">(</span><span class="n">fasta</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Tests the ASC germline set creation process using a fasta file or dictionary of alleles.</span>

<span class="sd">        Args:</span>
<span class="sd">            fasta (dict or path): A dictionary of allele sequences or a path to a fasta file containing alleles.</span>

<span class="sd">        Returns:</span>
<span class="sd">            defaultdict: A dictionary representing the ASC germline set.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">germline_set</span> <span class="o">=</span> <span class="n">create_allele_dict</span><span class="p">(</span><span class="n">fasta</span><span class="p">)</span>
    <span class="n">germline_set_asc</span> <span class="o">=</span> <span class="n">create_asc_germline_set</span><span class="p">(</span><span class="n">germline_set</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">germline_set_asc</span></div>


<div class="viewcode-block" id="profile">
<a class="viewcode-back" href="../../../GenAIRR.utilities.html#GenAIRR.utilities.asc_utilities.profile">[docs]</a>
<span class="k">def</span> <span class="nf">profile</span><span class="p">(</span><span class="n">fasta</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">       Profiles the performance of the ASC germline set creation process using a fasta file or dictionary of alleles.</span>

<span class="sd">       Args:</span>
<span class="sd">           fasta (dict or path): A dictionary of allele sequences or a path to a fasta file containing alleles.</span>
<span class="sd">   &quot;&quot;&quot;</span>
    <span class="kn">from</span> <span class="nn">cProfile</span> <span class="kn">import</span> <span class="n">Profile</span>
    <span class="n">germline_set</span> <span class="o">=</span> <span class="n">create_allele_dict</span><span class="p">(</span><span class="n">fasta</span><span class="p">)</span>
    <span class="n">cProfile</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="s1">&#39;create_asc_germline_set(germline_set)&#39;</span><span class="p">)</span>
    <span class="k">with</span> <span class="n">Profile</span><span class="p">()</span> <span class="k">as</span> <span class="n">profile</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">fib</span><span class="p">(</span><span class="mi">35</span><span class="p">)</span><span class="w"> </span><span class="si">= }</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="p">(</span>
            <span class="n">Stats</span><span class="p">(</span><span class="n">profile</span><span class="p">)</span>
            <span class="o">.</span><span class="n">strip_dirs</span><span class="p">()</span>
            <span class="o">.</span><span class="n">sort_stats</span><span class="p">(</span><span class="n">SortKey</span><span class="o">.</span><span class="n">CALLS</span><span class="p">)</span>
            <span class="o">.</span><span class="n">print_stats</span><span class="p">()</span>
        <span class="p">)</span></div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Thomas Konstantinovsky &amp; Ayelet Peres.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>