<!DOCTYPE html>
<html class="writer-html5" lang="python" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>GenAIRR.simulation.heavy_chain_sequence_augmentor &mdash; GenAIRR  documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=92fd9be5" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../../_static/documentation_options.js?v=d048f138"></script>
        <script src="../../../_static/doctools.js?v=888ff710"></script>
        <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            GenAIRR
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html">src</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">GenAIRR</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">GenAIRR.simulation.heavy_chain_sequence_augmentor</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for GenAIRR.simulation.heavy_chain_sequence_augmentor</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">pickle</span>
<span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">scipy.stats</span> <span class="k">as</span> <span class="nn">st</span>
<span class="kn">from</span> <span class="nn">..utilities</span> <span class="kn">import</span> <span class="n">AlleleNComparer</span>
<span class="kn">from</span> <span class="nn">..sequence</span> <span class="kn">import</span> <span class="n">HeavyChainSequence</span>
<span class="kn">from</span> <span class="nn">..simulation</span> <span class="kn">import</span> <span class="n">SequenceAugmentorArguments</span>
<span class="kn">from</span> <span class="nn">..simulation.sequence_augmentor_base</span> <span class="kn">import</span> <span class="n">SequenceAugmentorBase</span>
<span class="kn">from</span> <span class="nn">..utilities.data_config</span> <span class="kn">import</span> <span class="n">DataConfig</span>
<span class="kn">import</span> <span class="nn">base64</span>


<div class="viewcode-block" id="HeavyChainSequenceAugmentor">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.heavy_chain_sequence_augmentor.HeavyChainSequenceAugmentor">[docs]</a>
<span class="k">class</span> <span class="nc">HeavyChainSequenceAugmentor</span><span class="p">(</span><span class="n">SequenceAugmentorBase</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Augmentor class for simulating and augmenting heavy chain immunoglobulin sequences.</span>

<span class="sd">        This class extends the SequenceAugmentorBase to include functionalities specific to heavy chain sequences,</span>
<span class="sd">        such as handling V, D, and J gene segments, and applying heavy chain-specific corrections and augmentations.</span>

<span class="sd">        Attributes:</span>
<span class="sd">            alleles_used (list): List of gene segment types used in heavy chain sequences, typically including &#39;v&#39;, &#39;d&#39;, and &#39;j&#39;.</span>

<span class="sd">        Args:</span>
<span class="sd">            dataconfig (DataConfig): Configuration object containing allele information and simulation parameters.</span>
<span class="sd">            args (SequenceAugmentorArguments): Object containing arguments for sequence simulation and augmentation.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">alleles_used</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;v&#39;</span><span class="p">,</span> <span class="s1">&#39;d&#39;</span><span class="p">,</span> <span class="s1">&#39;j&#39;</span><span class="p">]</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">dataconfig</span><span class="p">:</span> <span class="n">DataConfig</span><span class="p">,</span> <span class="n">args</span><span class="p">:</span> <span class="n">SequenceAugmentorArguments</span> <span class="o">=</span> <span class="n">SequenceAugmentorArguments</span><span class="p">()):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">dataconfig</span><span class="p">,</span> <span class="n">args</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_add_distribution</span> <span class="o">=</span> <span class="n">st</span><span class="o">.</span><span class="n">beta</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_remove_distribution</span> <span class="o">=</span> <span class="n">st</span><span class="o">.</span><span class="n">beta</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_add_after_remove_distribution</span> <span class="o">=</span> <span class="n">st</span><span class="o">.</span><span class="n">beta</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">short_d_length</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">short_d_length</span>
        <span class="c1"># Class Misc</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">d_alleles</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataconfig</span><span class="o">.</span><span class="n">d_alleles</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataconfig</span><span class="o">.</span><span class="n">d_alleles</span><span class="p">[</span><span class="n">j</span><span class="p">]],</span>
                                <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">d_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">i</span><span class="o">.</span><span class="n">name</span><span class="p">:</span> <span class="n">i</span><span class="o">.</span><span class="n">ungapped_seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">d_alleles</span><span class="p">}</span>

        <span class="c1"># Loading Routines</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">load_correction_maps</span><span class="p">()</span>

    <span class="c1"># Loading Routines</span>
<div class="viewcode-block" id="HeavyChainSequenceAugmentor.load_correction_maps">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.heavy_chain_sequence_augmentor.HeavyChainSequenceAugmentor.load_correction_maps">[docs]</a>
    <span class="k">def</span> <span class="nf">load_correction_maps</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                Loads correction maps for V and D gene segments from the configuration object.</span>

<span class="sd">                These maps are used to adjust allele choices based on trimming events and to resolve ambiguities caused by similar alleles.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">importlib</span> <span class="kn">import</span> <span class="n">resources</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This will load the V start and V end maps that tell us given the amount removed from the start or end</span>
<span class="sd">        of a given allele what option are equally likely</span>
<span class="sd">        :return:</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">v_start_allele_correction_map</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataconfig</span><span class="o">.</span><span class="n">correction_maps</span><span class="p">[</span><span class="s1">&#39;V_5_TRIM_SIMILARITY_MAP&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">max_v_start_correction_map_value</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">v_start_allele_correction_map</span><span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">v_start_allele_correction_map</span><span class="p">)[</span><span class="mi">0</span><span class="p">]])</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">v_end_allele_correction_map</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataconfig</span><span class="o">.</span><span class="n">correction_maps</span><span class="p">[</span><span class="s1">&#39;V_3_TRIM_SIMILARITY_MAP&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">max_v_end_correction_map_value</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">v_end_allele_correction_map</span><span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">v_end_allele_correction_map</span><span class="p">)[</span><span class="mi">0</span><span class="p">]])</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">d_trim_correction_map</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataconfig</span><span class="o">.</span><span class="n">correction_maps</span><span class="p">[</span><span class="s1">&#39;D_5_3_TRIM_SIMILARITY_MAP&#39;</span><span class="p">]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">v_n_ambiguity_comparer</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataconfig</span><span class="o">.</span><span class="n">correction_maps</span><span class="p">[</span><span class="s1">&#39;V_N_AMBIGUITY_CORRECTION_GRAPH&#39;</span><span class="p">]</span></div>


<div class="viewcode-block" id="HeavyChainSequenceAugmentor.attribute_position">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.heavy_chain_sequence_augmentor.HeavyChainSequenceAugmentor.attribute_position">[docs]</a>
    <span class="k">def</span> <span class="nf">attribute_position</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">simulated</span><span class="p">,</span> <span class="n">position</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Determines the gene segment (V, D, J, or NP regions) associated with a given position in a simulated heavy chain sequence.</span>

<span class="sd">        Args:</span>
<span class="sd">            simulated (dict): Dictionary containing the simulated sequence and its metadata.</span>
<span class="sd">            position (int): The position within the sequence to be attributed.</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: The gene segment or region (&#39;v&#39;, &#39;d&#39;, &#39;j&#39;, &#39;np1&#39;, &#39;np2&#39;) associated with the given position.</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: If the position does not fall within the defined segments or regions.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_sequence_start&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">position</span> <span class="o">&lt;</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_sequence_end&#39;</span><span class="p">]:</span>
            <span class="k">return</span> <span class="s1">&#39;v&#39;</span>
        <span class="k">elif</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_sequence_end&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">position</span> <span class="o">&lt;</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;d_sequence_start&#39;</span><span class="p">]:</span>
            <span class="k">return</span> <span class="s1">&#39;np1&#39;</span>
        <span class="k">elif</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;d_sequence_start&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">position</span> <span class="o">&lt;</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;d_sequence_end&#39;</span><span class="p">]:</span>
            <span class="k">return</span> <span class="s1">&#39;d&#39;</span>
        <span class="k">elif</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;d_sequence_end&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">position</span> <span class="o">&lt;</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;j_sequence_start&#39;</span><span class="p">]:</span>
            <span class="k">return</span> <span class="s1">&#39;np2&#39;</span>
        <span class="k">elif</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;j_sequence_start&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">position</span> <span class="o">&lt;</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;j_sequence_end&#39;</span><span class="p">]:</span>
            <span class="k">return</span> <span class="s1">&#39;j&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Undefined Position!&#39;</span><span class="p">)</span></div>


<div class="viewcode-block" id="HeavyChainSequenceAugmentor.apply_deletion">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.heavy_chain_sequence_augmentor.HeavyChainSequenceAugmentor.apply_deletion">[docs]</a>
    <span class="k">def</span> <span class="nf">apply_deletion</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">simulated</span><span class="p">,</span> <span class="n">position</span><span class="p">):</span>
        <span class="k">pass</span>

        <span class="n">sequence</span> <span class="o">=</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">]</span>
        <span class="n">nucleotides_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
        <span class="n">region</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">attribute_position</span><span class="p">(</span><span class="n">simulated</span><span class="p">,</span> <span class="n">position</span><span class="p">)</span></div>


<div class="viewcode-block" id="HeavyChainSequenceAugmentor.correct_for_d_trims">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.heavy_chain_sequence_augmentor.HeavyChainSequenceAugmentor.correct_for_d_trims">[docs]</a>
    <span class="k">def</span> <span class="nf">correct_for_d_trims</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">simulated</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                Adjusts the D allele choices in the simulated sequence based on 5&#39; and 3&#39; trimming information.</span>

<span class="sd">                Args:</span>
<span class="sd">                    simulated (dict): Dictionary containing the simulated sequence and its metadata, including D allele trims.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Get the 5&#39; and 3&#39; trims of the d allele in the simulated sequence</span>
        <span class="n">trim_5</span> <span class="o">=</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;d_trim_5&#39;</span><span class="p">]</span>
        <span class="n">trim_3</span> <span class="o">=</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;d_trim_3&#39;</span><span class="p">]</span>
        <span class="c1"># infer the precalculated map what alleles should be the ground truth for this sequence based on the trim</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;d_allele&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">d_trim_correction_map</span><span class="p">[</span><span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;d_allele&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]][(</span><span class="n">trim_5</span><span class="p">,</span> <span class="n">trim_3</span><span class="p">)])</span></div>


<div class="viewcode-block" id="HeavyChainSequenceAugmentor.short_d_validation">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.heavy_chain_sequence_augmentor.HeavyChainSequenceAugmentor.short_d_validation">[docs]</a>
    <span class="k">def</span> <span class="nf">short_d_validation</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">simulated</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                Validates the length of the D gene segment in the simulated sequence. If the segment is shorter than a predefined threshold, it is labeled as &quot;Short-D&quot;.</span>

<span class="sd">                Args:</span>
<span class="sd">                    simulated (dict): Dictionary containing the simulated sequence and its metadata.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">d_length</span> <span class="o">=</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;d_sequence_end&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;d_sequence_start&#39;</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">d_length</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">short_d_length</span><span class="p">:</span>
            <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;d_allele&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Short-D&#39;</span><span class="p">]</span></div>


<div class="viewcode-block" id="HeavyChainSequenceAugmentor.fix_d_position_after_trimming_index_ambiguity">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.heavy_chain_sequence_augmentor.HeavyChainSequenceAugmentor.fix_d_position_after_trimming_index_ambiguity">[docs]</a>
    <span class="k">def</span> <span class="nf">fix_d_position_after_trimming_index_ambiguity</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">simulation</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                Corrects the start and end positions of the D gene segment in the simulated sequence to resolve ambiguities caused by trimming events.</span>

<span class="sd">                Args:</span>
<span class="sd">                    simulation (dict): Dictionary containing the simulated sequence and its metadata, including D allele positions and trimming details.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Extract Current D Metadata</span>
        <span class="n">d_start</span><span class="p">,</span> <span class="n">d_end</span> <span class="o">=</span> <span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;d_sequence_start&#39;</span><span class="p">],</span> <span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;d_sequence_end&#39;</span><span class="p">]</span>
        <span class="n">d_allele_remainder</span> <span class="o">=</span> <span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">][</span><span class="n">d_start</span><span class="p">:</span><span class="n">d_end</span><span class="p">]</span>
        <span class="n">d_allele_ref</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">d_dict</span><span class="p">[</span><span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;d_allele&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span>

        <span class="c1"># Get the junction inserted after trimming to the sequence</span>
        <span class="n">junction_5</span> <span class="o">=</span> <span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">][</span><span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;v_sequence_end&#39;</span><span class="p">]:</span><span class="n">d_start</span><span class="p">]</span>
        <span class="n">junction_3</span> <span class="o">=</span> <span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">][</span><span class="n">d_end</span><span class="p">:</span><span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;j_sequence_start&#39;</span><span class="p">]]</span>

        <span class="c1"># Get the trimming lengths</span>
        <span class="n">d_trim_5</span> <span class="o">=</span> <span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;d_trim_5&#39;</span><span class="p">]</span>
        <span class="n">d_trim_3</span> <span class="o">=</span> <span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;d_trim_3&#39;</span><span class="p">]</span>

        <span class="c1"># Get the trimmed off sections from the reference</span>
        <span class="n">trimmed_5</span> <span class="o">=</span> <span class="n">d_allele_ref</span><span class="p">[:</span><span class="n">d_trim_5</span><span class="p">]</span>
        <span class="n">trimmed_3</span> <span class="o">=</span> <span class="n">d_allele_ref</span><span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">d_allele_ref</span><span class="p">)</span> <span class="o">-</span> <span class="n">d_trim_3</span><span class="p">:]</span>

        <span class="c1"># check for overlap between generated junction and reference in the 5&#39; trim</span>
        <span class="k">for</span> <span class="n">a</span><span class="p">,</span> <span class="n">b</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">trimmed_5</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">junction_5</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]):</span>
            <span class="c1"># in case the current poistion in the junction matches the reference exapnd the d segment</span>
            <span class="k">if</span> <span class="n">a</span> <span class="o">==</span> <span class="n">b</span><span class="p">:</span>
                <span class="n">d_start</span> <span class="o">-=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>  <span class="c1"># if the continuous streak is broken or non-existent break!</span>
                <span class="k">break</span>

        <span class="c1"># check for overlap between generated junction and reference in the 3&#39; trim</span>
        <span class="k">for</span> <span class="n">a</span><span class="p">,</span> <span class="n">b</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">trimmed_3</span><span class="p">,</span> <span class="n">junction_3</span><span class="p">):</span>
            <span class="c1"># in case the current poistion in the junction matches the reference exapnd the d segment</span>
            <span class="k">if</span> <span class="n">a</span> <span class="o">==</span> <span class="n">b</span><span class="p">:</span>
                <span class="n">d_end</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>  <span class="c1"># if the continious streak is broken or non existant break!</span>
                <span class="k">break</span>

        <span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;d_sequence_start&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">d_start</span>
        <span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;d_sequence_end&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">d_end</span></div>


    <span class="c1"># Sequence Simulation</span>
<div class="viewcode-block" id="HeavyChainSequenceAugmentor.simulate_sequence">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.heavy_chain_sequence_augmentor.HeavyChainSequenceAugmentor.simulate_sequence">[docs]</a>
    <span class="k">def</span> <span class="nf">simulate_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                Simulates a heavy chain sequence, applying mutations and generating relevant metadata.</span>

<span class="sd">                Returns:</span>
<span class="sd">                    dict: A dictionary containing the simulated sequence, its metadata including gene segment positions, alleles, mutation rate, and trimming information.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">gen</span> <span class="o">=</span> <span class="n">HeavyChainSequence</span><span class="o">.</span><span class="n">create_random</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">dataconfig</span><span class="p">)</span>
        <span class="n">gen</span><span class="o">.</span><span class="n">mutate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutation_model</span><span class="p">)</span>

        <span class="n">data</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;sequence&quot;</span><span class="p">:</span> <span class="n">gen</span><span class="o">.</span><span class="n">mutated_seq</span><span class="p">,</span>
            <span class="s2">&quot;v_sequence_start&quot;</span><span class="p">:</span> <span class="n">gen</span><span class="o">.</span><span class="n">v_seq_start</span><span class="p">,</span>
            <span class="s2">&quot;v_sequence_end&quot;</span><span class="p">:</span> <span class="n">gen</span><span class="o">.</span><span class="n">v_seq_end</span><span class="p">,</span>
            <span class="s2">&quot;d_sequence_start&quot;</span><span class="p">:</span> <span class="n">gen</span><span class="o">.</span><span class="n">d_seq_start</span><span class="p">,</span>
            <span class="s2">&quot;d_sequence_end&quot;</span><span class="p">:</span> <span class="n">gen</span><span class="o">.</span><span class="n">d_seq_end</span><span class="p">,</span>
            <span class="s2">&quot;j_sequence_start&quot;</span><span class="p">:</span> <span class="n">gen</span><span class="o">.</span><span class="n">j_seq_start</span><span class="p">,</span>
            <span class="s2">&quot;j_sequence_end&quot;</span><span class="p">:</span> <span class="n">gen</span><span class="o">.</span><span class="n">j_seq_end</span><span class="p">,</span>
            <span class="s2">&quot;v_allele&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">gen</span><span class="o">.</span><span class="n">v_allele</span><span class="o">.</span><span class="n">name</span><span class="p">],</span>
            <span class="s2">&quot;d_allele&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">gen</span><span class="o">.</span><span class="n">d_allele</span><span class="o">.</span><span class="n">name</span><span class="p">],</span>
            <span class="s2">&quot;j_allele&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">gen</span><span class="o">.</span><span class="n">j_allele</span><span class="o">.</span><span class="n">name</span><span class="p">],</span>
            <span class="s1">&#39;mutation_rate&#39;</span><span class="p">:</span> <span class="n">gen</span><span class="o">.</span><span class="n">mutation_freq</span><span class="p">,</span>
            <span class="s1">&#39;v_trim_5&#39;</span><span class="p">:</span> <span class="n">gen</span><span class="o">.</span><span class="n">v_trim_5</span><span class="p">,</span>
            <span class="s1">&#39;v_trim_3&#39;</span><span class="p">:</span> <span class="n">gen</span><span class="o">.</span><span class="n">v_trim_3</span><span class="p">,</span>
            <span class="s1">&#39;d_trim_5&#39;</span><span class="p">:</span> <span class="n">gen</span><span class="o">.</span><span class="n">d_trim_5</span><span class="p">,</span>
            <span class="s1">&#39;d_trim_3&#39;</span><span class="p">:</span> <span class="n">gen</span><span class="o">.</span><span class="n">d_trim_3</span><span class="p">,</span>
            <span class="s1">&#39;j_trim_5&#39;</span><span class="p">:</span> <span class="n">gen</span><span class="o">.</span><span class="n">j_trim_5</span><span class="p">,</span>
            <span class="s1">&#39;j_trim_3&#39;</span><span class="p">:</span> <span class="n">gen</span><span class="o">.</span><span class="n">j_trim_3</span><span class="p">,</span>
            <span class="s1">&#39;corruption_event&#39;</span><span class="p">:</span> <span class="s1">&#39;no-corruption&#39;</span><span class="p">,</span>
            <span class="s1">&#39;corruption_add_amount&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s1">&#39;corruption_remove_amount&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s1">&#39;mutations&#39;</span><span class="p">:</span> <span class="p">{</span><span class="n">pos</span><span class="p">:</span> <span class="n">gen</span><span class="o">.</span><span class="n">mutations</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">gen</span><span class="o">.</span><span class="n">mutations</span><span class="p">)},</span>  <span class="c1"># sort the mutations by position</span>
            <span class="s2">&quot;Ns&quot;</span><span class="p">:</span> <span class="nb">dict</span><span class="p">(),</span>
        <span class="p">}</span>
        <span class="k">return</span> <span class="n">data</span></div>


<div class="viewcode-block" id="HeavyChainSequenceAugmentor.simulate_augmented_sequence">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.heavy_chain_sequence_augmentor.HeavyChainSequenceAugmentor.simulate_augmented_sequence">[docs]</a>
    <span class="k">def</span> <span class="nf">simulate_augmented_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                Generates an augmented heavy chain sequence, applying sequence corrections, potential corruption events at the beginning, and insertion of &#39;N&#39; bases.</span>

<span class="sd">                Returns:</span>
<span class="sd">                    dict: A dictionary containing the augmented sequence and associated metadata, ready for further analysis or training processes.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># 1. Simulate a Sequence from AIRRship</span>
        <span class="n">simulated</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">simulate_sequence</span><span class="p">()</span>

        <span class="c1"># 1.1 Correction - Correct Start/End Positions Based on Generated Junctions</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fix_v_position_after_trimming_index_ambiguity</span><span class="p">(</span><span class="n">simulated</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fix_d_position_after_trimming_index_ambiguity</span><span class="p">(</span><span class="n">simulated</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fix_j_position_after_trimming_index_ambiguity</span><span class="p">(</span><span class="n">simulated</span><span class="p">)</span>

        <span class="c1"># 1.2 Correction - Add All V Alleles That Cant be Distinguished Based on the Amount Cut from the V Allele</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">correct_for_v_end_cut</span><span class="p">(</span><span class="n">simulated</span><span class="p">)</span>

        <span class="c1"># 1.3 Correction - Add All D Alleles That Cant be Distinguished Based on the 5&#39; and 3&#39; Trims</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">correct_for_d_trims</span><span class="p">(</span><span class="n">simulated</span><span class="p">)</span>

        <span class="c1"># 2. Corrupt Begging of Sequence ( V Start )</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">perform_corruption</span><span class="p">():</span>
            <span class="c1"># Inside this method, based on the corruption event we will also adjust the respective ground truth v</span>
            <span class="c1"># alleles</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">corrupt_sequence_beginning</span><span class="p">(</span><span class="n">simulated</span><span class="p">)</span>

        <span class="c1"># 3. Add N&#39;s</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">insert_Ns</span><span class="p">(</span><span class="n">simulated</span><span class="p">)</span>

        <span class="c1"># 4. Adjust D Allele , if the simulated length is smaller than short_d_length</span>
        <span class="c1"># class property, change the d allele to the &quot;Short-D&quot; label</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">short_d_validation</span><span class="p">(</span><span class="n">simulated</span><span class="p">)</span>


        <span class="k">return</span> <span class="n">simulated</span></div>
</div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Thomas Konstantinovsky &amp; Ayelet Peres.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>