<!DOCTYPE html>
<html class="writer-html5" lang="python" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>GenAIRR.simulation.sequence_augmentor_base &mdash; GenAIRR  documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=92fd9be5" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../../_static/documentation_options.js?v=d048f138"></script>
        <script src="../../../_static/doctools.js?v=888ff710"></script>
        <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            GenAIRR
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html">src</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">GenAIRR</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">GenAIRR.simulation.sequence_augmentor_base</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for GenAIRR.simulation.sequence_augmentor_base</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">abc</span> <span class="kn">import</span> <span class="n">ABC</span><span class="p">,</span> <span class="n">abstractmethod</span>
<span class="kn">from</span> <span class="nn">..utilities.data_config</span> <span class="kn">import</span> <span class="n">DataConfig</span>
<span class="kn">from</span> <span class="nn">dataclasses</span> <span class="kn">import</span> <span class="n">dataclass</span>
<span class="kn">import</span> <span class="nn">enum</span>
<span class="kn">from</span> <span class="nn">..mutation</span> <span class="kn">import</span> <span class="n">MutationModel</span><span class="p">,</span> <span class="n">S5F</span>
<span class="kn">import</span> <span class="nn">scipy.stats</span> <span class="k">as</span> <span class="nn">st</span>
<span class="kn">import</span> <span class="nn">pickle</span>
<span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">base64</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>


<span class="c1"># Corruption Events</span>
<div class="viewcode-block" id="Event">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.Event">[docs]</a>
<span class="k">class</span> <span class="nc">Event</span><span class="p">(</span><span class="n">enum</span><span class="o">.</span><span class="n">Enum</span><span class="p">):</span>
    <span class="n">Remove</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="n">Add</span> <span class="o">=</span> <span class="mi">2</span>
    <span class="n">Remove_Before_Add</span> <span class="o">=</span> <span class="mi">3</span></div>



<span class="c1"># Arg Class</span>
<div class="viewcode-block" id="SequenceAugmentorArguments">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorArguments">[docs]</a>
<span class="nd">@dataclass</span>
<span class="k">class</span> <span class="nc">SequenceAugmentorArguments</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Configuration arguments for sequence augmentation processes, specifying mutation rates, indel simulation parameters, and other augmentation settings.</span>

<span class="sd">        Attributes:</span>
<span class="sd">            min_mutation_rate (float): The minimum mutation rate for simulating sequence mutations, defaulting to 0.003.</span>
<span class="sd">            max_mutation_rate (float): The maximum mutation rate for simulating sequence mutations, defaulting to 0.25.</span>
<span class="sd">            simulate_indels (bool): Flag to determine whether indels should be simulated, defaulting to False.</span>
<span class="sd">            max_indels (int): The maximum number of indels to simulate, defaulting to 5.</span>
<span class="sd">            deletion_proba (float): The probability of simulating a deletion event, defaulting to 0.5.</span>
<span class="sd">            insertion_proba (float): The probability of simulating an insertion event, defaulting to 0.5.</span>
<span class="sd">            n_ratio (float): The ratio of &#39;N&#39; bases to introduce as noise into sequences, defaulting to 0.02.</span>
<span class="sd">            max_sequence_length (int): The maximum length of sequences to simulate, defaulting to 512.</span>
<span class="sd">            mutation_model (MutationModel): The mutation model to use for sequence mutation simulation, defaulting to S5F.</span>
<span class="sd">            custom_mutation_model_path (str): The path to a custom mutation model file, if any, defaulting to None.</span>
<span class="sd">            nucleotide_add_coefficient (float): The coefficient for the distribution of nucleotides added, defaulting to 210.</span>
<span class="sd">            nucleotide_remove_coefficient (float): The coefficient for the distribution of nucleotides removed, defaulting to 310.</span>
<span class="sd">            nucleotide_add_after_remove_coefficient (float): The coefficient for the distribution of nucleotides added after removal, defaulting to 50.</span>
<span class="sd">            random_sequence_add_proba (float): The probability of adding a random sequence during augmentation, defaulting to 1.</span>
<span class="sd">            single_base_stream_proba (float): The probability of adding a single base stream during augmentation, defaulting to 0.</span>
<span class="sd">            duplicate_leading_proba (float): The probability of duplicating the leading base during augmentation, defaulting to 0.</span>
<span class="sd">            random_allele_proba (float): The probability of adding a random allele during augmentation, defaulting to 0.</span>
<span class="sd">            corrupt_proba (float): The probability of corrupting the sequence from the start, defaulting to 0.7.</span>
<span class="sd">            short_d_length (int): The minimum length required from the D allele to not be tagged as &quot;Short-D&quot;, defaulting to 5.</span>
<span class="sd">            kappa_lambda_ratio (float): The ratio of kappa to lambda light chains, defaulting to 0.5.</span>
<span class="sd">            save_mutations_record (bool): Whether to save the record of mutations in the sequence, defaulting to False.</span>
<span class="sd">            save_ns_record (bool): Whether to save the record of &#39;N&#39; bases in the sequence, defaulting to False.</span>
<span class="sd">        &quot;&quot;&quot;</span>
    <span class="n">min_mutation_rate</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.003</span>
    <span class="n">max_mutation_rate</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.25</span>
    <span class="n">simulate_indels</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">max_indels</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
    <span class="n">deletion_proba</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span>
    <span class="n">insertion_proba</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span>
    <span class="n">n_ratio</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.02</span>
    <span class="n">max_sequence_length</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">512</span>
    <span class="n">mutation_model</span><span class="p">:</span> <span class="n">MutationModel</span> <span class="o">=</span> <span class="n">S5F</span>
    <span class="n">custom_mutation_model_path</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">nucleotide_add_coefficient</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">210</span>
    <span class="n">nucleotide_remove_coefficient</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">310</span>
    <span class="n">nucleotide_add_after_remove_coefficient</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">50</span>
    <span class="n">random_sequence_add_proba</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="n">single_base_stream_proba</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">duplicate_leading_proba</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">random_allele_proba</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">corrupt_proba</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.7</span>
    <span class="n">short_d_length</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span>
    <span class="n">kappa_lambda_ratio</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.5</span>
    <span class="n">save_mutations_record</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="n">save_ns_record</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span></div>



<div class="viewcode-block" id="SequenceAugmentorBase">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorBase">[docs]</a>
<span class="k">class</span> <span class="nc">SequenceAugmentorBase</span><span class="p">(</span><span class="n">ABC</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">       An abstract base class for sequence augmentors, providing methods and attributes to simulate and augment immunoglobulin sequences.</span>

<span class="sd">       This class includes functionalities for simulating mutations, introducing noise (such as indels and base substitutions), and managing allele information for V, D, and J gene segments.</span>

<span class="sd">       Attributes:</span>
<span class="sd">           dataconfig (DataConfig): Configuration data for sequence simulation.</span>
<span class="sd">           min_mutation_rate (float): The minimum mutation rate for simulating sequence mutations.</span>
<span class="sd">           max_mutation_rate (float): The maximum mutation rate for simulating sequence mutations.</span>
<span class="sd">           mutation_model (MutationModel): The model used for simulating mutations within sequences.</span>
<span class="sd">           n_ratio (float): The ratio of &#39;N&#39; bases to introduce as noise into sequences.</span>
<span class="sd">           corrupt_proba (float): The probability of introducing corruption events at the beginning of sequences.</span>
<span class="sd">           simulate_indels (bool): Flag indicating whether to simulate insertions and deletions.</span>
<span class="sd">           max_indels (int): The maximum number of indels to simulate.</span>
<span class="sd">           deletion_proba (float): The probability of simulating a deletion event.</span>
<span class="sd">           insertion_proba (float): The probability of simulating an insertion event.</span>
<span class="sd">           save_mutations_record (bool): Flag indicating whether to save the record of mutations applied to sequences.</span>
<span class="sd">           save_ns_record (bool): Flag indicating whether to save the record of &#39;N&#39; bases introduced as noise.</span>

<span class="sd">       Args:</span>
<span class="sd">           dataconfig (DataConfig): The data configuration object containing settings for sequence simulation.</span>
<span class="sd">           args (SequenceAugmentorArguments): The arguments object containing simulation parameters.</span>
<span class="sd">       &quot;&quot;&quot;</span>
    <span class="n">alleles_used</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">dataconfig</span><span class="p">:</span> <span class="n">DataConfig</span><span class="p">,</span> <span class="n">args</span><span class="p">:</span> <span class="n">SequenceAugmentorArguments</span> <span class="o">=</span> <span class="n">SequenceAugmentorArguments</span><span class="p">()):</span>

        <span class="c1"># Sequence Generation Parameters</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">dataconfig</span> <span class="o">=</span> <span class="n">dataconfig</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">min_mutation_rate</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">min_mutation_rate</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">max_mutation_rate</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">max_mutation_rate</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">custom_mutation_model_path</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mutation_model</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">mutation_model</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">min_mutation_rate</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_mutation_rate</span><span class="p">,</span>
                                                      <span class="n">custom_model</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">custom_mutation_model_path</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mutation_model</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">mutation_model</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">min_mutation_rate</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_mutation_rate</span><span class="p">)</span>

        <span class="c1"># Noising Parameters</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">n_ratio</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">n_ratio</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">corrupt_proba</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">corrupt_proba</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_add_distribution</span> <span class="o">=</span> <span class="n">st</span><span class="o">.</span><span class="n">beta</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_remove_distribution</span> <span class="o">=</span> <span class="n">st</span><span class="o">.</span><span class="n">beta</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_add_after_remove_distribution</span> <span class="o">=</span> <span class="n">st</span><span class="o">.</span><span class="n">beta</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_add_coef</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">nucleotide_add_coefficient</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_remove_coef</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">nucleotide_remove_coefficient</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_add_after_remove_coef</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">nucleotide_add_after_remove_coefficient</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">random_sequence_add_proba</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">random_sequence_add_proba</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">single_base_stream_proba</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">single_base_stream_proba</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">duplicate_leading_proba</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">duplicate_leading_proba</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">random_allele_proba</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">random_allele_proba</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">simulate_indels</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">simulate_indels</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">max_indels</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">max_indels</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">deletion_proba</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">deletion_proba</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">insertion_proba</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">insertion_proba</span>
        <span class="c1"># Class Misc</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">chain_type</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">save_mutations_record</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">save_mutations_record</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">save_ns_record</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">save_ns_record</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">v_alleles</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataconfig</span><span class="o">.</span><span class="n">v_alleles</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataconfig</span><span class="o">.</span><span class="n">v_alleles</span><span class="p">[</span><span class="n">j</span><span class="p">]],</span>
                                <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">j_alleles</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataconfig</span><span class="o">.</span><span class="n">j_alleles</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataconfig</span><span class="o">.</span><span class="n">j_alleles</span><span class="p">[</span><span class="n">j</span><span class="p">]],</span>
                                <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">v_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">i</span><span class="o">.</span><span class="n">name</span><span class="p">:</span> <span class="n">i</span><span class="o">.</span><span class="n">ungapped_seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">v_alleles</span><span class="p">}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">j_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">i</span><span class="o">.</span><span class="n">name</span><span class="p">:</span> <span class="n">i</span><span class="o">.</span><span class="n">ungapped_seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">j_alleles</span><span class="p">}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">max_v_length</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">ungapped_seq</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">v_alleles</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">max_sequence_length</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">max_sequence_length</span>

        <span class="c1"># Loading Routines</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">load_correction_maps</span><span class="p">()</span>

    <span class="c1"># Loading Routines</span>
<div class="viewcode-block" id="SequenceAugmentorBase.load_correction_maps">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorBase.load_correction_maps">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">load_correction_maps</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            Abstract method to load allele correction maps, which are used to adjust allele choices based on observed mutations and indels.</span>

<span class="sd">            Implementations should load correction maps specific to the sequence type being augmented (e.g., heavy or light chain sequences).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">pass</span></div>


<div class="viewcode-block" id="SequenceAugmentorBase.get_original_index">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorBase.get_original_index">[docs]</a>
    <span class="k">def</span> <span class="nf">get_original_index</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">corrupted_index</span><span class="p">,</span> <span class="n">simulated</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculates the original index in the uncorrupted sequence corresponding to a given index in the corrupted sequence.</span>

<span class="sd">        Args:</span>
<span class="sd">            corrupted_index (int): The index in the corrupted sequence.</span>
<span class="sd">            simulated (dict): A dictionary containing simulation metadata, including details of corruption events and amounts.</span>

<span class="sd">        Returns:</span>
<span class="sd">            int or None: The original index in the uncorrupted sequence, or None if the index cannot be mapped back due to the nature of the corruption.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">corruption_event</span> <span class="o">=</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;corruption_event&#39;</span><span class="p">]</span>
        <span class="n">amount_added</span> <span class="o">=</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;corruption_add_amount&#39;</span><span class="p">]</span>
        <span class="n">amount_removed</span> <span class="o">=</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;corruption_remove_amount&#39;</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">corruption_event</span> <span class="o">==</span> <span class="s1">&#39;add&#39;</span><span class="p">:</span>
            <span class="c1"># If characters were added at the beginning, the original index is shifted to the right</span>
            <span class="k">return</span> <span class="n">corrupted_index</span> <span class="o">-</span> <span class="n">amount_added</span> <span class="k">if</span> <span class="n">corrupted_index</span> <span class="o">&gt;=</span> <span class="n">amount_added</span> <span class="k">else</span> <span class="kc">None</span>
        <span class="k">elif</span> <span class="n">corruption_event</span> <span class="o">==</span> <span class="s1">&#39;remove&#39;</span><span class="p">:</span>
            <span class="c1"># If characters were removed from the beginning, the original index is shifted to the left</span>
            <span class="k">return</span> <span class="n">corrupted_index</span> <span class="o">+</span> <span class="n">amount_removed</span>
        <span class="k">elif</span> <span class="n">corruption_event</span> <span class="o">==</span> <span class="s1">&#39;remove_before_add&#39;</span><span class="p">:</span>
            <span class="c1"># Combined effect of removal and addition</span>
            <span class="n">adjusted_index</span> <span class="o">=</span> <span class="n">corrupted_index</span> <span class="o">+</span> <span class="n">amount_removed</span>
            <span class="k">return</span> <span class="n">adjusted_index</span> <span class="o">-</span> <span class="n">amount_added</span> <span class="k">if</span> <span class="n">adjusted_index</span> <span class="o">&gt;=</span> <span class="n">amount_added</span> <span class="k">else</span> <span class="kc">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># If no corruption or unknown corruption event, assume the index is unchanged</span>
            <span class="k">return</span> <span class="n">corrupted_index</span></div>


<div class="viewcode-block" id="SequenceAugmentorBase.get_allele_spesific_n_positions">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorBase.get_allele_spesific_n_positions">[docs]</a>
    <span class="k">def</span> <span class="nf">get_allele_spesific_n_positions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">simulated</span><span class="p">,</span> <span class="n">n_positions</span><span class="p">,</span> <span class="n">allele</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                Introduces &#39;N&#39; bases into the simulated sequence based on the configured &#39;N&#39; ratio, simulating sequencing errors or low-quality bases.</span>

<span class="sd">                Args:</span>
<span class="sd">                    simulated (dict): A dictionary containing the simulated sequence and metadata.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">n_positions</span> <span class="k">if</span>
                <span class="n">simulated</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">allele</span><span class="si">}</span><span class="s1">_sequence_start&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">i</span> <span class="o">&lt;=</span> <span class="n">simulated</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">allele</span><span class="si">}</span><span class="s1">_sequence_end&#39;</span><span class="p">]]</span></div>


<div class="viewcode-block" id="SequenceAugmentorBase.attribute_position">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorBase.attribute_position">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">attribute_position</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">simulated</span><span class="p">,</span> <span class="n">position</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method will take a simulated sequence and a position and return the name of the region it is in</span>
<span class="sd">        :param simulated:</span>
<span class="sd">        :param position:</span>
<span class="sd">        :return:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">pass</span></div>


    <span class="c1"># Noise Introducing Methods</span>
<div class="viewcode-block" id="SequenceAugmentorBase.insert_Ns">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorBase.insert_Ns">[docs]</a>
    <span class="k">def</span> <span class="nf">insert_Ns</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">simulated</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Introduces &#39;N&#39; bases into the simulated sequence based on the configured &#39;N&#39; ratio, simulating sequencing errors or low-quality bases.</span>

<span class="sd">                Args:</span>
<span class="sd">                    simulated (dict): A dictionary containing the simulated sequence and metadata.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">sequence</span> <span class="o">=</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">]</span>
        <span class="c1"># Calculate how many Ns we should insert</span>
        <span class="n">num_replacements</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_ratio</span><span class="p">)</span>
        <span class="n">nucleotides_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_replacements</span><span class="p">):</span>
            <span class="c1"># Get random position in the sequence</span>
            <span class="n">index</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">nucleotides_list</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
            <span class="c1"># Log the N insertion event in the sequence metadata</span>
            <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;Ns&#39;</span><span class="p">][</span><span class="n">index</span><span class="p">]</span> <span class="o">=</span> <span class="n">nucleotides_list</span><span class="p">[</span><span class="n">index</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;&gt;&#39;</span> <span class="o">+</span> <span class="s1">&#39;N&#39;</span>
            <span class="c1"># Make the N insertion</span>
            <span class="n">nucleotides_list</span><span class="p">[</span><span class="n">index</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;N&quot;</span>

        <span class="c1"># Concatenate the list back into a string</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">nucleotides_list</span><span class="p">)</span>

        <span class="c1"># Sort the N&#39;s insertion log</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;Ns&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="n">pos</span><span class="p">:</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;Ns&#39;</span><span class="p">][</span><span class="n">pos</span><span class="p">]</span> <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;Ns&#39;</span><span class="p">])}</span>

        <span class="c1"># Get All the N Poistion Inserted to the V Allele</span>
        <span class="n">v_allele_n_positions</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_allele_spesific_n_positions</span><span class="p">(</span><span class="n">simulated</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;Ns&#39;</span><span class="p">]),</span> <span class="s1">&#39;v&#39;</span><span class="p">)</span>
        <span class="c1"># Get Original Positions For Correction</span>
        <span class="n">v_allele_n_positions_original</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">get_original_index</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">simulated</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">v_allele_n_positions</span><span class="p">]</span>

        <span class="n">indistinguishable_v_alleles</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">v_n_ambiguity_comparer</span><span class="o">.</span><span class="n">find_indistinguishable_alleles</span><span class="p">(</span>
            <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_allele&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">v_allele_n_positions_original</span><span class="p">)</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_allele&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_allele&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span>
            <span class="nb">set</span><span class="p">(</span><span class="n">indistinguishable_v_alleles</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_allele&#39;</span><span class="p">]))</span>

        <span class="c1"># Adjust mutation rate to account for added &quot;N&#39;s&quot; by adding the ratio of N&#39;s added to the mutation rate</span>
        <span class="c1"># because we randomly sample position we might get a ratio of N&#39;s less than what was defined by the n_ratio</span>
        <span class="c1"># property, thus we recalculate the actual inserted ration of N&#39;s</span>
        <span class="c1"># Also make sure we only count the N&#39;s the were inserted in the sequence and not in the added slack</span>
        <span class="c1"># as we might consider all the slack as noise in general</span>
        <span class="n">Ns_in_pure_sequence</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;Ns&#39;</span><span class="p">]</span> <span class="k">if</span>
                               <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_sequence_start&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">i</span> <span class="o">&lt;=</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;j_sequence_end&#39;</span><span class="p">]]</span>
        <span class="n">pure_sequence_length</span> <span class="o">=</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;j_sequence_end&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_sequence_start&#39;</span><span class="p">]</span>

        <span class="c1"># exception handling</span>
        <span class="k">if</span> <span class="n">pure_sequence_length</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># handle division by zero</span>
            <span class="n">simulated_n_ratio</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">simulated_n_ratio</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">Ns_in_pure_sequence</span><span class="p">)</span> <span class="o">/</span> <span class="n">pure_sequence_length</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;mutation_rate&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="n">simulated_n_ratio</span></div>


<div class="viewcode-block" id="SequenceAugmentorBase.remove_event">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorBase.remove_event">[docs]</a>
    <span class="k">def</span> <span class="nf">remove_event</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">simulated</span><span class="p">,</span> <span class="n">autocorrect</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Simulates the removal of a portion of the sequence from the beginning, adjusting metadata accordingly.</span>

<span class="sd">                Args:</span>
<span class="sd">                    simulated (dict): A dictionary containing the simulated sequence and metadata.</span>
<span class="sd">                    autocorrect (bool): Whether to automatically correct allele choices based on the removal event.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Update Simulation Metadata</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;corruption_event&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;remove&#39;</span>
        <span class="c1"># Sample how much to remove</span>
        <span class="n">v_length</span> <span class="o">=</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_sequence_end&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_sequence_start&#39;</span><span class="p">]</span>
        <span class="n">amount_to_remove</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sample_nucleotide_remove_distribution</span><span class="p">(</span><span class="n">v_length</span><span class="p">)</span>
        <span class="c1"># remove from the start of the sequence the sampled amount</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">][</span><span class="n">amount_to_remove</span><span class="p">:]</span>
        <span class="c1"># Update Simulation Metadata</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;corruption_remove_amount&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">amount_to_remove</span>
        <span class="c1"># Update mutation log, remove the mutations that were removed with the remove event</span>
        <span class="c1"># and while updating the mutations log correct the position of the mutations accordingly</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;mutations&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="n">i</span> <span class="o">-</span> <span class="n">amount_to_remove</span><span class="p">:</span> <span class="n">j</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;mutations&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span>
                                  <span class="n">i</span> <span class="o">&gt;</span> <span class="n">amount_to_remove</span><span class="p">}</span>
        <span class="c1"># Adjust mutation rate</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">correct_mutation_rate</span><span class="p">(</span><span class="n">simulated</span><span class="p">)</span>

        <span class="c1"># Adjust Start/End Position Accordingly</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_sequence_start&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_sequence_end&#39;</span><span class="p">]</span> <span class="o">-=</span> <span class="n">amount_to_remove</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;d_sequence_start&#39;</span><span class="p">]</span> <span class="o">-=</span> <span class="n">amount_to_remove</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;d_sequence_end&#39;</span><span class="p">]</span> <span class="o">-=</span> <span class="n">amount_to_remove</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;j_sequence_start&#39;</span><span class="p">]</span> <span class="o">-=</span> <span class="n">amount_to_remove</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;j_sequence_end&#39;</span><span class="p">]</span> <span class="o">-=</span> <span class="n">amount_to_remove</span>

        <span class="c1"># Correction - Add All V Alleles That Cant be Distinguished Based on the Amount Cut from the V Allele</span>
        <span class="k">if</span> <span class="n">autocorrect</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">correct_for_v_start_cut</span><span class="p">(</span><span class="n">simulated</span><span class="p">)</span></div>


<div class="viewcode-block" id="SequenceAugmentorBase.add_event">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorBase.add_event">[docs]</a>
    <span class="k">def</span> <span class="nf">add_event</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">simulated</span><span class="p">,</span> <span class="n">amount</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Simulates the addition of bases to the beginning of the sequence, adjusting metadata accordingly.</span>

<span class="sd">                Args:</span>
<span class="sd">                    simulated (dict): A dictionary containing the simulated sequence and metadata.</span>
<span class="sd">                    amount (int, optional): The number of bases to add. If not specified, the amount is sampled from a configured distribution.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Update Simulation Metadata</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;corruption_event&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;add&#39;</span>
        <span class="c1"># Sample the Amount to Add by default, if a spesific value is given via the amount variable,use it.</span>
        <span class="n">amount_to_add</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sample_nucleotide_add_distribution</span><span class="p">()</span> <span class="k">if</span> <span class="n">amount</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">amount</span>
        <span class="c1"># Sample the method by which addition will be made</span>
        <span class="n">method</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sample_corruption_add_method</span><span class="p">()</span>
        <span class="c1"># Modify the sequence</span>
        <span class="n">modified_sequence</span> <span class="o">=</span> <span class="n">method</span><span class="p">(</span><span class="n">amount_to_add</span><span class="p">,</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">])</span>
        <span class="c1"># Validate the modified sequence, make sure we didn&#39;t over add pass our max sequence size</span>
        <span class="n">modified_sequence</span><span class="p">,</span> <span class="n">amount_to_add</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">validate_sequence_length_after_addition</span><span class="p">(</span><span class="n">modified_sequence</span><span class="p">,</span>
                                                                                        <span class="n">amount_to_add</span><span class="p">)</span>
        <span class="c1"># Update Simulation Sequence</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">modified_sequence</span>

        <span class="c1"># Update Simulation Metadata</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;corruption_add_amount&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">amount_to_add</span>
        <span class="c1"># Update mutation log positions</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;mutations&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="n">i</span> <span class="o">+</span> <span class="n">amount_to_add</span><span class="p">:</span> <span class="n">j</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;mutations&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>

        <span class="c1"># Adjust Start/End Position Accordingly</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_sequence_start&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="n">amount_to_add</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_sequence_end&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="n">amount_to_add</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;d_sequence_start&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="n">amount_to_add</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;d_sequence_end&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="n">amount_to_add</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;j_sequence_start&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="n">amount_to_add</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;j_sequence_end&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="n">amount_to_add</span></div>


<div class="viewcode-block" id="SequenceAugmentorBase.remove_before_add_event">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorBase.remove_before_add_event">[docs]</a>
    <span class="k">def</span> <span class="nf">remove_before_add_event</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">simulated</span><span class="p">):</span>
        <span class="c1"># ----- REMOVE PART -----#</span>
        <span class="c1"># Sample how much to remove</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">remove_event</span><span class="p">(</span><span class="n">simulated</span><span class="p">,</span> <span class="n">autocorrect</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>  <span class="c1"># Dont Correct For Removed Section YET! First Lets Add</span>

        <span class="c1"># ----- ADD PART -----#</span>
        <span class="c1"># Sample how much to add after removal occurred</span>
        <span class="n">amount_to_add</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sample_nucleotide_add_after_remove_distribution</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">add_event</span><span class="p">(</span><span class="n">simulated</span><span class="p">,</span> <span class="n">amount</span><span class="o">=</span><span class="n">amount_to_add</span><span class="p">)</span>

        <span class="c1"># Update Simulation Metadata</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;corruption_event&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;remove_before_add&#39;</span>

        <span class="c1"># Check If The Addition Recreated Some of the Removed Section by Chance</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">correct_for_v_start_add</span><span class="p">(</span><span class="n">simulated</span><span class="p">)</span>
        <span class="c1"># Correct for the removed section of the V now that we have check for the reconstruction by chance</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">correct_for_v_start_cut</span><span class="p">(</span><span class="n">simulated</span><span class="p">)</span></div>


<div class="viewcode-block" id="SequenceAugmentorBase.corrupt_sequence_beginning">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorBase.corrupt_sequence_beginning">[docs]</a>
    <span class="k">def</span> <span class="nf">corrupt_sequence_beginning</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">simulated</span><span class="p">):</span>
        <span class="c1"># Sample Corruption Event</span>
        <span class="n">event</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_random_event</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">event</span> <span class="ow">is</span> <span class="n">Event</span><span class="o">.</span><span class="n">Remove</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">remove_event</span><span class="p">(</span><span class="n">simulated</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">event</span> <span class="ow">is</span> <span class="n">Event</span><span class="o">.</span><span class="n">Add</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">add_event</span><span class="p">(</span><span class="n">simulated</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">event</span> <span class="ow">is</span> <span class="n">Event</span><span class="o">.</span><span class="n">Remove_Before_Add</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">remove_before_add_event</span><span class="p">(</span><span class="n">simulated</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Unknown Corruption Event </span><span class="si">{</span><span class="n">event</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span></div>


    <span class="c1"># Different V Start &quot;Add&quot; Event Scenarios</span>
<div class="viewcode-block" id="SequenceAugmentorBase.random_nucleotides">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorBase.random_nucleotides">[docs]</a>
    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">random_nucleotides</span><span class="p">(</span><span class="n">amount</span><span class="p">,</span> <span class="n">sequence</span><span class="p">):</span>
        <span class="n">random_seq</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">choices</span><span class="p">([</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">],</span> <span class="n">k</span><span class="o">=</span><span class="n">amount</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">random_seq</span> <span class="o">+</span> <span class="n">sequence</span></div>


<div class="viewcode-block" id="SequenceAugmentorBase.duplicate_leading">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorBase.duplicate_leading">[docs]</a>
    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">duplicate_leading</span><span class="p">(</span><span class="n">amount</span><span class="p">,</span> <span class="n">sequence</span><span class="p">):</span>
        <span class="n">cap</span> <span class="o">=</span> <span class="n">amount</span> <span class="k">if</span> <span class="n">amount</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="k">else</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>
        <span class="k">return</span> <span class="n">sequence</span><span class="p">[:</span><span class="n">cap</span><span class="p">]</span> <span class="o">+</span> <span class="n">sequence</span></div>


<div class="viewcode-block" id="SequenceAugmentorBase.random_allele_section">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorBase.random_allele_section">[docs]</a>
    <span class="k">def</span> <span class="nf">random_allele_section</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">amount</span><span class="p">,</span> <span class="n">sequence</span><span class="p">):</span>
        <span class="n">random_allele</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">v_alleles</span><span class="p">)</span><span class="o">.</span><span class="n">ungapped_seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
        <span class="n">cap</span> <span class="o">=</span> <span class="n">amount</span> <span class="k">if</span> <span class="n">amount</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">random_allele</span><span class="p">)</span> <span class="k">else</span> <span class="nb">len</span><span class="p">(</span><span class="n">random_allele</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>
        <span class="k">return</span> <span class="n">random_allele</span><span class="p">[:</span><span class="n">cap</span><span class="p">]</span> <span class="o">+</span> <span class="n">sequence</span></div>


<div class="viewcode-block" id="SequenceAugmentorBase.single_base_stream">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorBase.single_base_stream">[docs]</a>
    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">single_base_stream</span><span class="p">(</span><span class="n">amount</span><span class="p">,</span> <span class="n">sequence</span><span class="p">):</span>
        <span class="n">random_base</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">([</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;N&#39;</span><span class="p">])</span> <span class="o">*</span> <span class="n">amount</span>
        <span class="k">return</span> <span class="n">random_base</span> <span class="o">+</span> <span class="n">sequence</span></div>


    <span class="c1"># Correction Functions</span>
<div class="viewcode-block" id="SequenceAugmentorBase.correct_for_v_end_cut">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorBase.correct_for_v_end_cut">[docs]</a>
    <span class="k">def</span> <span class="nf">correct_for_v_end_cut</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">simulated</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Corrects the V allele choices based on the trimming of the sequence&#39;s end, ensuring the chosen alleles remain consistent with the modified sequence.</span>

<span class="sd">                Args:</span>
<span class="sd">                    simulated (dict): A dictionary containing the simulated sequence and metadata, including the current V allele and its trimmed length.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">equivalent_alleles</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">v_end_allele_correction_map</span><span class="p">[</span><span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_allele&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]][</span>
            <span class="nb">min</span><span class="p">(</span><span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_sequence_end&#39;</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_v_end_correction_map_value</span><span class="p">)]</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_allele&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_allele&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">equivalent_alleles</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_allele&#39;</span><span class="p">]))</span></div>


<div class="viewcode-block" id="SequenceAugmentorBase.correct_for_v_start_cut">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorBase.correct_for_v_start_cut">[docs]</a>
    <span class="k">def</span> <span class="nf">correct_for_v_start_cut</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">simulated</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">       Adjusts the V allele choices in response to the sequence&#39;s start being trimmed, maintaining consistency between the chosen alleles and the altered sequence.</span>

<span class="sd">               Args:</span>
<span class="sd">                   simulated (dict): A dictionary containing the simulated sequence and metadata, particularly details about the V allele and the amount trimmed from the start.</span>
<span class="sd">       &quot;&quot;&quot;</span>
        <span class="n">removed</span> <span class="o">=</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;corruption_remove_amount&#39;</span><span class="p">]</span>
        <span class="n">equivalent_alleles</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">v_start_allele_correction_map</span><span class="p">[</span><span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_allele&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]][</span>
            <span class="nb">min</span><span class="p">(</span><span class="n">removed</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_v_start_correction_map_value</span><span class="p">)]</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_allele&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_allele&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">equivalent_alleles</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_allele&#39;</span><span class="p">]))</span></div>


<div class="viewcode-block" id="SequenceAugmentorBase.correct_for_v_start_add">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorBase.correct_for_v_start_add">[docs]</a>
    <span class="k">def</span> <span class="nf">correct_for_v_start_add</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">simulated</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Evaluates and corrects the V allele start position after an addition event, ensuring the V allele start position remains accurate despite sequence modifications.</span>

<span class="sd">        Args:</span>
<span class="sd">            simulated (dict): A dictionary containing the simulated sequence and metadata, including information about addition events and their impact on the V allele start position.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;corruption_event&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;remove_before_add&#39;</span><span class="p">:</span>
            <span class="n">amount_added</span> <span class="o">=</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;corruption_add_amount&#39;</span><span class="p">]</span>
            <span class="n">amount_removed</span> <span class="o">=</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;corruption_remove_amount&#39;</span><span class="p">]</span>

            <span class="n">add_section</span> <span class="o">=</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">][:</span><span class="n">amount_added</span><span class="p">]</span>
            <span class="n">removed_section</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">v_dict</span><span class="p">[</span><span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_allele&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]][:</span><span class="n">amount_removed</span><span class="p">]</span>

            <span class="n">min_length</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">add_section</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">removed_section</span><span class="p">))</span>
            <span class="n">to_adjust</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">min_length</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
                <span class="c1"># Compare characters from the end</span>
                <span class="k">if</span> <span class="n">add_section</span><span class="p">[</span><span class="o">-</span><span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="n">removed_section</span><span class="p">[</span><span class="o">-</span><span class="n">i</span><span class="p">]:</span>
                    <span class="n">to_adjust</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>  <span class="c1"># Mismatch found, halt the iteration</span>
                    <span class="k">break</span>

            <span class="c1"># Adjust V Start</span>
            <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_sequence_start&#39;</span><span class="p">]</span> <span class="o">-=</span> <span class="n">to_adjust</span>
            <span class="c1"># Adjust Removed Amount</span>
            <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;corruption_remove_amount&#39;</span><span class="p">]</span> <span class="o">-=</span> <span class="n">to_adjust</span></div>


<div class="viewcode-block" id="SequenceAugmentorBase.correct_mutation_rate">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorBase.correct_mutation_rate">[docs]</a>
    <span class="k">def</span> <span class="nf">correct_mutation_rate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">simulated</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Recalculates and updates the mutation rate in the simulated sequence metadata to reflect the actual mutation rate after sequence modifications such as insertions or deletions.</span>

<span class="sd">        Args:</span>
<span class="sd">            simulated (dict): A dictionary containing the simulated sequence and metadata, including the current mutation log and sequence length.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># recalculate the mutation ratio based on the current mutation log</span>
        <span class="n">n_mutations</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;mutations&#39;</span><span class="p">])</span>
        <span class="c1"># the actual sequence length without any noise at the start or at the end</span>
        <span class="n">sequence_length</span> <span class="o">=</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;j_sequence_end&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;v_sequence_start&#39;</span><span class="p">]</span>
        <span class="n">mutation_ratio</span> <span class="o">=</span> <span class="n">n_mutations</span> <span class="o">/</span> <span class="n">sequence_length</span>
        <span class="c1"># update the mutation_rate in the simulation log</span>
        <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;mutation_rate&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">mutation_ratio</span></div>


<div class="viewcode-block" id="SequenceAugmentorBase.validate_sequence_length_after_addition">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorBase.validate_sequence_length_after_addition">[docs]</a>
    <span class="k">def</span> <span class="nf">validate_sequence_length_after_addition</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sequence</span><span class="p">,</span> <span class="n">added</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                Ensures that the sequence length does not exceed the maximum allowed length after an addition event, trimming excess bases if necessary.</span>

<span class="sd">                Args:</span>
<span class="sd">                    sequence (str): The sequence after the addition event.</span>
<span class="sd">                    added (int): The number of bases added to the sequence.</span>

<span class="sd">                Returns:</span>
<span class="sd">                    tuple: A tuple containing the validated sequence and the adjusted number of bases added.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_sequence_length</span><span class="p">:</span>
            <span class="c1"># Calculate how much did we over add to based on our maximum sequence length</span>
            <span class="n">slack</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_sequence_length</span>
            <span class="c1"># remove the slack from the begging and update the added variable to match the final added amount</span>
            <span class="k">return</span> <span class="n">sequence</span><span class="p">[</span><span class="n">slack</span><span class="p">:],</span> <span class="n">added</span> <span class="o">-</span> <span class="n">slack</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">sequence</span><span class="p">,</span> <span class="n">added</span></div>


<div class="viewcode-block" id="SequenceAugmentorBase.fix_v_position_after_trimming_index_ambiguity">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorBase.fix_v_position_after_trimming_index_ambiguity">[docs]</a>
    <span class="k">def</span> <span class="nf">fix_v_position_after_trimming_index_ambiguity</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">simulation</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                Resolves any ambiguities in V allele positions resulting from sequence trimming, ensuring accurate representation of V allele boundaries.</span>

<span class="sd">                Args:</span>
<span class="sd">                    simulation (dict): A dictionary containing the simulated sequence and metadata, including V allele positions and trimming details.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Extract Current V Metadata</span>
        <span class="n">v_start</span><span class="p">,</span> <span class="n">v_end</span> <span class="o">=</span> <span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;v_sequence_start&#39;</span><span class="p">],</span> <span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;v_sequence_end&#39;</span><span class="p">]</span>
        <span class="n">v_allele_remainder</span> <span class="o">=</span> <span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">][</span><span class="n">v_start</span><span class="p">:</span><span class="n">v_end</span><span class="p">]</span>
        <span class="n">v_allele_ref</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">v_dict</span><span class="p">[</span><span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;v_allele&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span>

        <span class="c1"># Get the junction inserted after trimming to the sequence</span>
        <span class="n">junction_3</span> <span class="o">=</span> <span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">][</span><span class="n">v_end</span><span class="p">:</span><span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;d_sequence_start&#39;</span><span class="p">]]</span>

        <span class="c1"># Get the trimming lengths</span>
        <span class="n">v_trim_3</span> <span class="o">=</span> <span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;v_trim_3&#39;</span><span class="p">]</span>

        <span class="c1"># Get the trimmed off sections from the reference</span>
        <span class="n">trimmed_3</span> <span class="o">=</span> <span class="n">v_allele_ref</span><span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">v_allele_ref</span><span class="p">)</span> <span class="o">-</span> <span class="n">v_trim_3</span><span class="p">:]</span>

        <span class="c1"># check for overlap between generated junction and reference in the 3&#39; trim</span>
        <span class="k">for</span> <span class="n">a</span><span class="p">,</span> <span class="n">b</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">trimmed_3</span><span class="p">,</span> <span class="n">junction_3</span><span class="p">):</span>
            <span class="c1"># in case the current poistion in the junction matches the reference exapnd the d segment</span>
            <span class="k">if</span> <span class="n">a</span> <span class="o">==</span> <span class="n">b</span><span class="p">:</span>
                <span class="n">v_end</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>  <span class="c1"># if the continuous streak is broken or non-existent break!</span>
                <span class="k">break</span>

        <span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;v_sequence_end&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">v_end</span></div>


<div class="viewcode-block" id="SequenceAugmentorBase.fix_j_position_after_trimming_index_ambiguity">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorBase.fix_j_position_after_trimming_index_ambiguity">[docs]</a>
    <span class="k">def</span> <span class="nf">fix_j_position_after_trimming_index_ambiguity</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">simulation</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            Addresses ambiguities in J allele positions due to sequence trimming, ensuring J allele boundaries are accurately represented.</span>

<span class="sd">                Args:</span>
<span class="sd">                    simulation (dict): A dictionary containing the simulated sequence and metadata, including J allele positions and trimming details.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Extract Current J Metadata</span>
        <span class="n">j_start</span><span class="p">,</span> <span class="n">j_end</span> <span class="o">=</span> <span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;j_sequence_start&#39;</span><span class="p">],</span> <span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;j_sequence_end&#39;</span><span class="p">]</span>
        <span class="n">j_allele_remainder</span> <span class="o">=</span> <span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">][</span><span class="n">j_start</span><span class="p">:</span><span class="n">j_end</span><span class="p">]</span>
        <span class="n">j_allele_ref</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">j_dict</span><span class="p">[</span><span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;j_allele&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span>

        <span class="c1"># Get the junction inserted after trimming to the sequence</span>
        <span class="n">junction_5</span> <span class="o">=</span> <span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">][</span><span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;d_sequence_end&#39;</span><span class="p">]:</span><span class="n">j_start</span><span class="p">]</span>

        <span class="c1"># Get the trimming lengths</span>
        <span class="n">j_trim_5</span> <span class="o">=</span> <span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;j_trim_5&#39;</span><span class="p">]</span>

        <span class="c1"># Get the trimmed off sections from the reference</span>
        <span class="n">trimmed_5</span> <span class="o">=</span> <span class="n">j_allele_ref</span><span class="p">[:</span><span class="n">j_trim_5</span><span class="p">]</span>

        <span class="c1"># check for overlap between generated junction and reference in the 5&#39; trim</span>
        <span class="k">for</span> <span class="n">a</span><span class="p">,</span> <span class="n">b</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">trimmed_5</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">junction_5</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]):</span>
            <span class="c1"># in case the current poistion in the junction matches the reference exapnd the d segment</span>
            <span class="k">if</span> <span class="n">a</span> <span class="o">==</span> <span class="n">b</span><span class="p">:</span>
                <span class="n">j_start</span> <span class="o">-=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>  <span class="c1"># if the continuous streak is broken or non-existent break!</span>
                <span class="k">break</span>

        <span class="n">simulation</span><span class="p">[</span><span class="s1">&#39;j_sequence_start&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">j_start</span></div>


    <span class="c1"># Sequence Simulation</span>
<div class="viewcode-block" id="SequenceAugmentorBase.simulate_sequence">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorBase.simulate_sequence">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">simulate_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Abstract method to simulate an immunoglobulin sequence, incorporating V(D)J recombination, mutations, and potentially other forms of sequence variation.</span>

<span class="sd">        Implementations should generate a simulated sequence along with relevant metadata, such as the positions of gene segments and mutations.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: A dictionary containing the simulated sequence and associated metadata.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">pass</span></div>


    <span class="c1"># Interface</span>
<div class="viewcode-block" id="SequenceAugmentorBase.process_before_return">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorBase.process_before_return">[docs]</a>
    <span class="k">def</span> <span class="nf">process_before_return</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">simulated</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Performs final processing on the simulated sequence before returning it, including formatting allele lists and encoding mutation and &#39;N&#39; base records.</span>

<span class="sd">        Args:</span>
<span class="sd">            simulated (dict): A dictionary containing the simulated sequence and metadata.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Convert Allele From List to Comma Seperated String</span>
        <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alleles_used</span><span class="p">:</span>
            <span class="n">simulated</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s1">_allele&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">simulated</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s1">_allele&#39;</span><span class="p">])</span>

        <span class="c1"># convert mutations and N log to base64 for proper tabular preservation</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">save_ns_record</span><span class="p">:</span>
            <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;Ns&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">base64</span><span class="o">.</span><span class="n">b64encode</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;Ns&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span><span class="s1">&#39;ascii&#39;</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">simulated</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;Ns&#39;</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">save_mutations_record</span><span class="p">:</span>
            <span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;mutations&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">base64</span><span class="o">.</span><span class="n">b64encode</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">simulated</span><span class="p">[</span><span class="s1">&#39;mutations&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span><span class="s1">&#39;ascii&#39;</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">simulated</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;mutations&#39;</span><span class="p">)</span></div>


<div class="viewcode-block" id="SequenceAugmentorBase.simulate_augmented_sequence">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorBase.simulate_augmented_sequence">[docs]</a>
    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">simulate_augmented_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                Abstract method to simulate and augment an immunoglobulin sequence, incorporating additional forms of sequence variation beyond basic V(D)J recombination and mutations.</span>

<span class="sd">                Implementations might include more complex forms of noise or variation, such as indels or base modifications, and should adjust the sequence and metadata accordingly.</span>

<span class="sd">                Returns:</span>
<span class="sd">                    dict: A dictionary containing the augmented sequence and associated metadata.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">pass</span></div>


    <span class="c1"># Misc Methods</span>
    <span class="k">def</span> <span class="nf">_sample_nucleotide_add_distribution</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            Samples from a predefined distribution to determine the number of nucleotides to add during a sequence augmentation event.</span>

<span class="sd">                Returns:</span>
<span class="sd">                    int: The number of nucleotides to be added to the sequence.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">sample</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_add_coef</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_add_distribution</span><span class="o">.</span><span class="n">rvs</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">sample</span><span class="o">.</span><span class="n">item</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">_sample_nucleotide_remove_distribution</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">v_length</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                Samples from a predefined distribution to decide the number of nucleotides to remove, with a maximum limit based on the V segment length.</span>

<span class="sd">                Args:</span>
<span class="sd">                    v_length (int): The length of the V segment in the sequence.</span>

<span class="sd">                Returns:</span>
<span class="sd">                    int: The number of nucleotides to be removed from the sequence.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Sample amount based on predefined distribution</span>
        <span class="n">sample</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_remove_coef</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_remove_distribution</span><span class="o">.</span><span class="n">rvs</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span><span class="o">.</span><span class="n">item</span><span class="p">()</span>
        <span class="c1"># make sure that no matter how much we get from sampling the predefined distribution we wont get a value</span>
        <span class="c1"># larger than the total v length we have in our sequence</span>
        <span class="n">sample</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">sample</span><span class="p">,</span> <span class="n">v_length</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">sample</span>

    <span class="k">def</span> <span class="nf">_sample_nucleotide_add_after_remove_distribution</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                Samples from a predefined distribution to determine the number of nucleotides to add after a removal event during sequence augmentation.</span>

<span class="sd">                Returns:</span>
<span class="sd">                    int: The number of nucleotides to be added to the sequence following a removal event.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">sample</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_add_after_remove_coef</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">nucleotide_add_after_remove_distribution</span><span class="o">.</span><span class="n">rvs</span><span class="p">(</span>
            <span class="n">size</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">sample</span><span class="o">.</span><span class="n">item</span><span class="p">()</span>

<div class="viewcode-block" id="SequenceAugmentorBase.perform_corruption">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorBase.perform_corruption">[docs]</a>
    <span class="k">def</span> <span class="nf">perform_corruption</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                Determines whether a corruption event should be performed on the sequence based on the configured probability.</span>

<span class="sd">                Returns:</span>
<span class="sd">                    bool: True if a corruption event should be performed, False otherwise.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">bool</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">binomial</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">corrupt_proba</span><span class="p">))</span></div>


<div class="viewcode-block" id="SequenceAugmentorBase.sample_random_event">
<a class="viewcode-back" href="../../../GenAIRR.simulation.html#GenAIRR.simulation.sequence_augmentor_base.SequenceAugmentorBase.sample_random_event">[docs]</a>
    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">sample_random_event</span><span class="p">():</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                Randomly selects a corruption event type (e.g., add, remove, remove before add) based on predefined probabilities.</span>

<span class="sd">                Returns:</span>
<span class="sd">                    Event: An enumerated value representing the selected corruption event type.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">([</span><span class="n">Event</span><span class="o">.</span><span class="n">Remove</span><span class="p">,</span> <span class="n">Event</span><span class="o">.</span><span class="n">Add</span><span class="p">,</span> <span class="n">Event</span><span class="o">.</span><span class="n">Remove_Before_Add</span><span class="p">],</span> <span class="n">size</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="p">[</span><span class="mf">0.4</span><span class="p">,</span> <span class="mf">0.4</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">])</span><span class="o">.</span><span class="n">item</span><span class="p">()</span></div>


    <span class="k">def</span> <span class="nf">_sample_corruption_add_method</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">               Randomly selects a method for adding nucleotides to the sequence during an augmentation event based on predefined probabilities.</span>

<span class="sd">               Returns:</span>
<span class="sd">                   function: A reference to the function that will be used to add nucleotides to the sequence.</span>
<span class="sd">       &quot;&quot;&quot;</span>
        <span class="n">method</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choices</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">random_nucleotides</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">single_base_stream</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">duplicate_leading</span><span class="p">,</span>
                                 <span class="bp">self</span><span class="o">.</span><span class="n">random_allele_section</span><span class="p">],</span>
                                <span class="n">weights</span><span class="o">=</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">random_sequence_add_proba</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">single_base_stream_proba</span><span class="p">,</span>
                                         <span class="bp">self</span><span class="o">.</span><span class="n">duplicate_leading_proba</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">random_allele_proba</span><span class="p">],</span> <span class="n">k</span><span class="o">=</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">method</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">columns</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">               Returns a list of column names that would be present in the output of the simulate_augmented_sequence method.</span>

<span class="sd">               Returns:</span>
<span class="sd">                   list: A list of strings representing the column names.</span>
<span class="sd">       &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">simulate_augmented_sequence</span><span class="p">())</span></div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Thomas Konstantinovsky &amp; Ayelet Peres.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>